[영화 극장 선택] 페토 1, 2단계 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트
- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?
-

## 코드 리뷰 커뮤니케이션

#### 안녕하세요 두루!

지난 오목 미션에서 두루덕에 정말 많은 깨달음과 성장을 했는데 
또 만나게 되어 너무 반갑습니다 ☺️

작성된 코드 중 데이터 바인딩을 사용한 부분에서 이미지를 바인딩 할 때 
BindingAdapter를 적용한 곳도 있고 적용하지 않은 부분도 있는데요 !

시간이 부족해 디테일하게 적용하지 못하고 바인딩을 통해 View에 대한 참조를 얻는 정도로만 사용했습니다. 🥲

기존의 findViewById를 사용할 때 매번 레이아웃 트리에 대한 탐색 비용을 제거할 수 있단 점은 정말 신세계였지만
데이터 바인딩을 사용하면서 데이터 바인딩을 사용함에 있어서 이걸 왜 사용해야하지? 에 대한 의문이 크게 든 몇가지 요인이 있습니다. 😞

첫 번째로, XML에 로직이 들어가는게 굉장히 어색하게 느껴졌어요 🧐
아마 데이터 바인딩을 사용하는 이유 중 하나로 Activity나 다른 뷰에 존재해야할 코드들이 줄어든다는 장점이 있을 수도 있지만
결국 이 코드들도 XML에 존재하고 이렇게 코드가 분산되는게 저는 영 맘에 들지 않았답니다.

두 번째로, BindingAdapter 같은 불필요한 코드들이 너무 많아진다고 생각해요
그냥 `setImageResource` 같이 기존에 제공하는 메서드들을 사용하면 되는데 이걸 꼭 만들어야할까 ?..

개인적으로 BindingAdapter를 이해하는데 있어 많은 러닝 커브를 느꼈고 이에 대한 학습 비용으로 인해
미션에 차질이 생길 것 같아  초기엔 BindingAdapter를 만들고 적용하다 점점 포기하게 되었습니다 🥹

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

https://github.com/user-attachments/assets/5a93e13f-f47a-4d74-86be-f49bab7fa227

woowahan-pjs committed:
Initial commit
# Conflicts:
#	README.md

woowahan-pjs committed:
feat: set up the project

woowahan-pjs committed:
feat: set up the project

chanho0908 committed:
[영화 티켓 예매 1, 2단계] 페토 미션 제출합니다 (#105)

* docs : 1단계 기능 요구사항 명세서 추가

* feat : 영화 아이템 레이아웃 추가

* feat : 영화 목록 한 개 화면 출력 구현

* feat : 예약 완료 화면 UI 구현

* feat : 예약 완료 화면 이동 구현

* docs : 2단계 기능 요구 사항 추가

* feat : 영화 예매 레이아웃 추가

* feat : 영화 상영 및 예매 가능 날짜 상태 클래스 추가

* feat : 주말, 평일 상태 클래스 추가

* feat : 영화 상영 가능 시간 객체 추가

* feat : 영화 목록 -> 예매 화면 이동 완료

* refactor : 영화 목록 아이템 레이아웃 수정

* refactor : 상영 일자 데이터 클래스로 수정

* refactor : Movie 상영 일자 프로퍼티 객체로 수정

* refactor : 문자열 조합 함수 추가

* feat : Movie 상영 일자 연월일 분리하여 전달

* feat : 상영 날짜, 예매 시간 스피너 추가

* feat : 인원수 증감 버튼 기능 추가

* chore : ktlint 포맷 수정

* refactor : 상영 날짜 출력 형식 포맷 수정

* feat : 문자열 포맷팅 클래스 추가

* feat : 예매자 타입 상태 클래스 추가

* feat : 예매 인원수 원시값 매핑 및 예매 금액 계산 함수 추가

* feat : 예매 확인 버튼 클릭시 다이얼로그 출력 기능 추가

* feat : 예매 완료 후 예매 완료 화면에 예매 정보 전달 기능 추가

* feat : 영화 목록의 상영 날짜 포맷팅 수정

* chore : 전체 파일 패키지 수정

* build : espresso test dependency add

* feat : 화면 전환시 데이터 유지 기능 추가

* docs : 리드미 체크 추가

* feat : 스피너 어댑터 리스너 구현체 분리

* feat : 자정을 나타내는 상수 가시성 변경자 수정

* build : espresso device dependency gradle setting 제거

* test : 화면 회전 테스트 수정

* test : 안드로이드 테스트 DisplayName 제거

* test : 인원 감소 테스트시 기존 인원 검증 로직 추가

* test : given,when,then 주석 추가

* refactor : movie item orientation 추가

* test : 테스트시 불필요한 화면 화전 제거

* test : 테스트 코드 패키지 수정

* refactor : static import 수정

* feat : 영화 List Adapter에 ViewHolder Pattern 적용

* refactor : 천원 단위 콤파 포맷팅 수정

* refactor : 예매 화면 스크롤뷰 추가

* refactor : 다이얼로그 버튼 텍스트 리소스를 직접 참조하도록 수정

* refactor : 영화 도메인 모델의 포스터 이미지 프로퍼티 타입 수정

* feat : serializable Intent API 33 대응 확장함수 추가

* feat : 영화, 상영일, 상영 시간 Class 직렬화 적용

* feat : 인원수 증감 함수, 예약 가능한 최소 인원  검증을 위한 함수 추가

* feat : 예약 정보 객체 추가

* feat : 예약 정보 객체를 만드는 함수 분리

* feat : 리스트 어댑터 추상 클래스 추가

* feat : 영화 리스트 어댑터 추상 클래스 적용

* refactor : 테스트 코드 패키지 수정

* feat : 영화 일급 컬랙션 추가 및 더미 데이터 캐싱 추가

* feat : 예약 가능한 시간이 없을 때 안내 메시지 출력 기능 추가

* refactor : 예약 domain Model의 예약 시간, 날짜 Type을 도메인에 맞게 수정

* refactor : 예약 인원수 객체의 예약 가능 최소 인원 검증 로직 수정

* refactor : 예매 화면에서 영화의 정보만 스크롤 되도록 수정

* feat : 예매 시간 객체 함수에 사용되는 인자를 프로퍼티로 수정

* refactor : 예약 가능한 시간을 받아올시 함수 중복 호출 수정

---------

Co-authored-by: ijh1298 <imjh0417@gmail.com>

chanho0908 committed:
rename : 영화 목록 패키지명 수정

chanho0908 committed:
refactor : 러닝타임의 "분" 포맷을 strings resource로 분리 및 적용

chanho0908 committed:
feat : 포스터 리소스 객체 분리

chanho0908 committed:
refactor : Movie 객체 패키징 수정

chanho0908 committed:
build : mockk gradle dependecy 추가

chanho0908 committed:
test : Movie Fixture 추가

chanho0908 committed:
refactor : viewHolder 데이터 바인딩 함수에 수신 객체 지정 람다 with 적용

chanho0908 committed:
feat : 영화 목록 화면 MVP Pattern 적용 및 mockk 테스트 추가

- 테스트용 MovieModel 추가

chanho0908 committed:
rename : 예약 정보 도메인 패키징 수정

chanho0908 committed:
rename : 영화 목록 Presenter 패키지 수정

chanho0908 committed:
feat : 예매 화면 기본 UI 로직 MVP 적용

- 영화 정보 가져오기
- 인원수 증감

chanho0908 committed:
feat : 예약 완료 후 예약 내역 화면 이동 로직 MVP 적용 완료

chanho0908 committed:
refactor : 영화 정보를 예매 화면에 전달할 때 domain Model를 사용하지 않도록 수정

chanho0908 committed:
feat : 이미지가 없을 경우 placeholder 이미지를 보여주도록 구현

chanho0908 committed:
feat : 영화 상열일 출력 로직 MVP 적용 완료

chanho0908 committed:
feat : 영화 상영 시간 출력 로직 MVP 적용

chanho0908 committed:
feat : 날짜 선택시 시간 초기화 MVP 적용

chanho0908 committed:
feat : 예매 인원수 복원 기능 MVP 적용

chanho0908 committed:
file : 영화 포스터, 광고 이미지 추가

chanho0908 committed:
chore : 영화 일급 컬랙션, 포스터 이미지 객체 제거

chanho0908 committed:
feat : 영화 정보를 가져오는 데이터 소스 추가

chanho0908 committed:
feat : 영화 객체 Id 값 추가

chanho0908 committed:
feat : 광고 객체 추가

chanho0908 committed:
build : recyclerview dependency 추가

chanho0908 committed:
feat : 파일 네임으로 Drawble Resource Id를 가져오는 확장함수 추가

chanho0908 committed:
feat : 광고 뷰홀더, 레이아웃 추가

chanho0908 committed:
feat : 리사이클러뷰 & (광고, 영화) 멀티 뷰홀더 적용 및 UI Model 추가

chanho0908 committed:
feat : 영화 목록 화면 Presenter 기능 추가

- Domain Model -> Ui Model 매핑 함수 추가
- 10000개의 영화 데이터를 가져오는 로직 추가

chanho0908 committed:
refactor : domain -> UiModel 매핑 확장함수 위치 수정

chanho0908 committed:
feat : 예약 화면 Movie UIModel 적용 완료

chanho0908 committed:
refactor : 영화 저장소 객체 게터 오퍼레이터 함수 추가, 사이즈 반환 함수 제거

chanho0908 committed:
feat : espresso recyclerview test dependency 추가

chanho0908 committed:
feat : 영화 리사이클러뷰 화면 테스트 추가

chanho0908 committed:
docs : 기능 요구사항 업데이트

chanho0908 committed:
feat : 영화 객체의 ID를 영화 리스트 생성시 Index 값으로 수정

chanho0908 committed:
test : 영화 저장소 객체 생성시 테스트 누락 수정

chanho0908 committed:
refactor : 예약 프로퍼티 중 티켓 타입 제거

chanho0908 committed:
feat : 좌석 선택 layout 추가 및 intent 전달 테스트 완료

chanho0908 committed:
feat : 좌석 화면 뒤로가기 기능 추가

chanho0908 committed:
feat : 예매 좌석 도메인 모델 추가

chanho0908 committed:
feat : 테이블 레이아웃의 테로 사용될 좌표 객체 추가

chanho0908 committed:
feat : 좌선 "선택" 로직 MVP 적용

chanho0908 committed:
feat : 좌석 등급별 가격 정책 클래스 추가

chanho0908 committed:
feat : 좌석 선택시 예매 가능한 인원수 초과시 토스트 출력 기능 추가

- 좌표, 좌석 픽스처 추가

chanho0908 committed:
feat : 예매한 좌석 가격 출력 기능 추가

chanho0908 committed:
feat : 좌석을 예약 인원만큼 선택하지 않았을 때 버튼 비활성화 기능 추가

chanho0908 committed:
feat : 예약 완료시 예약 정보 전달 및 예약 완료 화면에서 예약 정보 출력 기능 추가

chanho0908 committed:
refactor : Presenter, Contract 패키지 분리

chanho0908 committed:
test : 불필요한 테스트 제거, 영화 더미 데이터 수정으로 인한 테스트 수정

chanho0908 committed:
refactor : 영화 목록화면에서 예매 화면 전환 로직 수정

chanho0908 committed:
rename : 좌석 화면 테스트 패키지 수정

chanho0908 committed:
chore : 불필요한 테스트 제거

chanho0908 committed:
feat : 좌석 선택후 예매 완료 버튼 클릭시 다이얼로그 출력 추가

chanho0908 committed:
refactor : 예매 화면에서 좌석 선택 화면 이동 전환 로직 수정

chanho0908 committed:
refactor : 좌석 선택 화면에서 예약 완료 화면으로 이동 로직 수정

chanho0908 committed:
refactor : 광고 출력 후 리사이클러뷰의 인덱스가 한개씩 밀려서 다음 영화가 출력되는 현상 수정

chanho0908 committed:
refactor : 영화 전체 리스트 반환시 새로운 컬렉션을 깊은 복사해 반환하도록 수정

chanho0908 committed:
feat : 영화관 좌석수 만큼 예매 인원 제한 기능 추가

chanho0908 committed:
refactor : 좌석 선택 화면 레이아웃 수정

chanho0908 committed:
feat : 문서용 디렉토리 추가

chanho0908 committed:
refactor : 리사이클러뷰의 리스너 인덱스 반환으로 인한 문제 해결 및 영화 데이터 접근 최적화

chanho0908 committed:
chore : 기타 코드 개선

- 예약 완료 화면 패키지 분리
- 좌석 테이블 레이아웃의 셀 높이가 0으로 되어 좌석이 보이지 않던 문제 수정
- 사용하지 않는 AbstractListViewAdapter class 제거

chanho0908 committed:
refactor : Presenter의 DIP 위반 수정

chanho0908 committed:
refactor : Presenter의 비즈니스 로직 제거

- view가 초기화 됐을 때 0원을 보여주기 위한 호출 추가

chanho0908 committed:
refactor : Presenter의 UI 로직 제거

chanho0908 committed:
refactor : 테이블 레이아웃 코드 개선

chanho0908 committed:
refactor : Prestener에서 비즈니스 로직을 요청 하지 않는 메서드 제거

chanho0908 committed:
refactor :  Presenter 함수 네이밍 수정

chanho0908 committed:
refactor : 좌석 화면에서 intent로 전달 받은 기본 예약 정보를 Presenter가 관리하도록 수정

chanho0908 committed:
feat : 예약 완료 화면 MVP 적용

chanho0908 committed:
refactor : 좌석 예매 -> 예매 완료 화면에 티켓 전달시 좌석 리스트를 전달하도록 수정

chanho0908 committed:
rename : presenter, contract 패키지 수정

chanho0908 committed:
refactor : 광고 레이아웃 수정

chanho0908 committed:
feat : 영화관 객체 추가 및 영화관별 상영 시간 객체 추가

chanho0908 committed:
feat : 영화관 목록 Layout, ViewHolder, Adapter 추가

chanho0908 committed:
feat : 영화와 영화관 더미 데이터 추가

chanho0908 committed:
feat : 영화관 목록 바텀 시트 추가

chanho0908 committed:
feat : 영화 선택시 영화관 목록 바텀시트 출력 기능 추가

chanho0908 committed:
feat : 예매 화면 전환시 상영 시간 정보 전달 기능 추가

chanho0908 committed:
feat : 상영 기간 객체에서 전체 상영 기간을 받아 시작일과 종료일을 판단하도록 수정

chanho0908 committed:
feat : 영화 상영 기간 출력 기능 추가

chanho0908 committed:
refactor : 영화, 영화관 더미 데이터 수정

chanho0908 committed:
feat : 동일한 날짜는 1개만 반환하도록 수정

chanho0908 committed:
feat : 예매 시간이 당일일때, 아닐 때 예매 가능한 시간 반환 함수 추가

chanho0908 committed:
feat : 예매 날짜에 따른 시간 스피너 출력 기능 추가

chanho0908 committed:
test : 영화 리스트 화면 -> 예매 화면 전환 테스트

doabletuple committed:
feat : 예매 완료시 극장 이름 출력 기능 추가

doabletuple committed:
feat : 영화 선택 화면 data binding 적용

doabletuple committed:
feat : 영화관 바텀 시트 data binding 적용

doabletuple committed:
feat : 영화 예매 화면 data binding 적용

doabletuple committed:
feat : 좌석 선택 화면 data binding 적용

doabletuple committed:
feat : 예매 완료 화면 data binding 적용

doabletuple committed:
refactor : Data binding 누락된 코드 수정

doabletuple committed:
feat : 바텀 네비게이션 추가

doabletuple committed:
feat : 기본 Fragment 파일 추가

doabletuple committed:
feat : 홈 화면 Fragment 적용

chanho0908 committed:
test : 테스트 코드 수정

chanho0908 committed:
Squashed commit of the following:

commit fdbabbca2e9d73946d489928c290fc7a9f8721d2
Author: DoableTuple <doabletuple@gmail.com>
Date:   Thu May 1 13:30:30 2025 +0900

    refactor : 사용되지 않는 Activity 제거

chanho0908 committed:
refactor : recyclerView LayoutManager 지정 로직을 xml 속성으로 수정

chanho0908 committed:
chore : 불필요한 클래스 제거

chanho0908 committed:
refactor : 상영 기간 클래스의 상영 시작일, 종료을 프로퍼티에 custom getter 추가

chanho0908 committed:
refactor : 예매 가능 시간 판별 여부 함수의 조건을 변수로 분리

chanho0908 committed:
refactor : 영화 목록 Presenter의 영화관 저장소 프로퍼티를 필드로 수정

chanho0908 committed:
feat : 도메인 모델과 중복되는 UI용 좌표 객체 제거

chanho0908 committed:
refactor : 예약 화면의 복구 로직 onRestoreInstanceState로 수정

chanho0908 committed:
refactor : 극장 뷰홀더 전체에 클릭 리스너 지정

chanho0908 committed:
refactor : 예약 화면 바인딩 미적용 수정, Presenter 초기화시 인원수를 보여주도록 수정

chanho0908 committed:
refactor : 좌석 추가,삭제 상태 판별 로직 수정

chanho0908 committed:
refactor : interface와 구현체의 메서드 인자명이 다른 것들 수정

chanho0908 committed:
build : Fragment gradle Dependency add

chanho0908 committed:
refactor : 앱 최초 시작시 영화 목록 화면이 보이도록 수정

chanho0908 committed:
refactor : 영화 목록 뷰홀더 리스너 바인딩 위치 수정

chanho0908 committed:
test : 예약 완료 화면 극장명 출력 테스트 추가, 패키지 분리

chanho0908 committed:
chore : 앱 전체 패키지 수정

chanho0908 committed:
refactor : mock 객체는 한번만 초기화 되도록 수정

chanho0908 committed:
chore : 불필요한 프래그먼트 코드 제거

chanho0908 committed:
refactor : 예약 완료 화면 Presenter 초기화시 예약 정보를 렌더링 하도록 수정

chanho0908 committed:
feat : 예약 가능한 시간이 없는 극장 필터링 기능 추가

chanho0908 committed:
refactor : 영화관 더미 데이터 수정

chanho0908 committed:
test : 테스트 함수명 오류 수정

chanho0908 committed:
chore : test code given when then 순서 수정

chanho0908 committed:
refactor : xml namespace 오타 수정

chanho0908 committed:
refactor : 포맷팅 함수 중복 호출 제거

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -41,6 +42,9 @@ android {
             excludes += "win32-x86*/**"
         }
     }
+    dataBinding {
+        enable = true
+    }
```
살짝 작은 팁을 한가지 남겨볼게요~
gradle 도 Kotlin 이니까 아래처럼도 작성할 수 있습니다!
~~~kotlin
dataBinding.enable = true
~~~

chanho0908 replied:
오호라 !! 이런 것도 DSL이라고 하는걸까요 ? 코틀린 정말 매력이 철철 넘치네요 ~~ 🤣

Gyuil-Hwnag replied:
이전에 groovy 때도 여러 기능도 있긴 했었는데, 저희가 익숙한 Kotlin 기반이다 보니 이게 될까? 라고 생각했던 것들 중에서 되는 것들이 은근 있었어요ㅎㅎ😀

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/booking/BookingActivityTest.kt:
```diff
@@ -0,0 +1,109 @@
+class BookingActivityTest {
+    private lateinit var scenario: ActivityScenario<BookingActivity>
+
+    @Before
+    fun setUp() {
```
테스트 코드 너무 잘 작성해주셨네요👍
참고차 남겨두었습니다~
@Before 이라는 어노테이션이 있는만큼, @After 라는 어노테이션도 있습니다!
추가로 @Before / BeforeEach 이렇게 존재하는데 두개는 어떤 차이점을 가지고 있을까요~🤔

chanho0908 replied:
`Before`와 `BeforeEach`는 모두 테스트간의 독립성을 위해 **매 테스트마다 초기화** 되어야 하는 요소를 초기화하기 위해 사용합니다.

예를 들어 10번의 테스트를 해야한다면 10번의 모두 실행됩니다.

JVM 환경에서 실행되는 **유닛 테스트**는 JUnit5를 사용하는 경우 `@BeforeEach`를 사용할 수 있고,

**Android Instrumentation 테스트**처럼 JUnit4 기반인 경우에는 `@Before`를 사용해야 합니다.

다만 `@Before`를 사용하는 경우, 매번 액티비티를 실행해야 하므로 테스트 실행 비용이 커질 수 있습니다.

그래서 `@BeforeAll, @BeforeClass` 를 사용하면 단 한번만 초기화할 수 있습니다.

하지만 인원 증감 버튼 테스트처럼 액티비티 상태 초기화가 필요한 상황에서는 성능보다 테스트 독립성을 우선시하는 **불가피한 트레이드오프**라고 생각됩니다!

Gyuil-Hwnag replied:
오호 페토의 의견 아주 좋습니다~
해당 내용에 대해서 한번 정리를 해둔다면, 추후 테스트 코드를 작성할 때 인스턴스 관리등 여러가지 측면에서 잘 이해해 볼 수 있겠다는 생각이 들어서 남겨둔 내용이었어요!

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/booking/BookingActivityTest.kt:
```diff
@@ -0,0 +1,109 @@
+
+    @Test
+    fun `인원_증가_버튼을_누르면_인원이_1_증가한다`() {
+        // given
+        onView(withId(R.id.tv_people_count)).check(matches(withText("1")))
+
+        // when
+        onView(withId(R.id.btn_increase)).perform(click())
+
+        // then
+        onView(withId(R.id.tv_people_count)).check(matches(withText("2")))
```
테스트 코드에서 아래와 같은 함수를 추가함으로써, 반복적으로 들어가는 내용들에 대해서 줄여볼 수 있겠다는 생각이 들었어요💪
아래와 같은 내용처럼 테스트 함수용 Extension를 몇개 만들어 볼 수 있겠군요~
~~~kotlin
fun ViewInteraction.performClick() = this.perform(click())
~~~

Gyuil-Hwnag replied:
당장에 생각나는 것들은 matches, isDisplayed, click 등 이렇게가 있을 수 있겠군요!

chanho0908 replied:
테스트 코드도 확장 함수를 활용할 수 있군요 !!

개인적으로 이런 확장 함수를 정말 좋아하는데 크루들의 반발이 심해서 마음이 아마요 두루 🥲🥲🥲

다들 `Util Class`같다고 하시는데 전 `Util Class`와는 전혀 다르다고 생각하거든요 !

top-level함수나 Object 의 정적 메서드들은 경계가 없는 그냥 아무대나 둥둥 떠다니는 아이라
객체지향적 관점에서 봤을 때 지양해야 한다는 의견은 너무나 동의합니다.

하지만 확장 함수는 수신 객체의 멤버는 아니지만 멤버인 것 처럼 활동하는, 경계가 있는 아이라고 생각합니다.

또한 개방 폐쇄 원칙에 아주 잘 맞다고 생각해요 ! 

실제 객체를 변경시키지 않으면서, 말 그대로 확장하고 있으니까요 !

Gyuil-Hwnag replied:
오호 그렇군요~
크루분들하고 여러 이야기를 나눴었나 보군요

해당 내용을 남겨둘 때는 아래와 같은 이유였어요
`이렇게 작성을 하는 방법도 있으며, 이점을 참고해 보면 좋을 것 같다`
해당 내용에서 출발해서 객체지향적 관점까지 가다니... 그렇다면 한가지 한번 이야기를 해보면 좋을게 생각이 났어요

디자인 패턴 / 아키텍쳐 / 여러 원칙 및 규칙등은 무엇을 위해서 나온것 일까요?
개인적으로는 이 두가지 키워드 (`확장성 / 생산성`) 로부터 출발하지 않았을까 싶어요.

그치만, 지금 크루분들이 나누는 얘기는 규칙과 원칙을 지키다 보니 오히려  중요한 출발점인 `확장성 / 생산성` 보다 이 두가지 키워드를 잘 해보기 위해서 나온 규칙을 더욱 중시한다는 뭔가... 우선순위가 살짝 바뀌게 되었다는 생각도 살짝 들기는 하네요
(물론 이건 개인적인 의견이라서 반대되는 입장 언제든 환영입니다😀)

chanho0908 replied:
옛날 옛적 얘기를 할 때 단군 할아버지가 나오시는데
저희 한태 단군 할어버지는 엘런 튜링이 아닐까 싶어요

앨런 튜링 할아버지 때 부터 지금까지 프로그래밍이란 기술이 발전해오며
수 많은 시행착오를 겪고 동일한 문제들을 겪었을 겁니다

그리고 이런 문제를 해결하는 방법들을 획일화하고 최적화 한 것이 말씀하신
디자인 패턴이나 아키텍처가 탄생한 계기인 것 같습니다

그리고 이 방법들 조차 또 다른 문제들을 겪으며 점점 발전해 나가는 것 같아요 ! ☺️

저도 두루의 의견에 동의합니다 :)

제가 아직 경험이 적어 두루의 뜻을 모두 이해기엔 부족하지만
규칙과 원칙 또한 특정한 무언가를 더 잘 만들기 위한 일종의 "수단과 방법"인 것 같습니다

그 수단과 방법이 옳다고만은 할 수 없는 것 같아요 정답은 없으니까요 !

대신 개선하고자 하는 것, 목표하는 것을 정확히 이루기 위해 더 그 수단과 방법을
더 잘 사용하는 것을 고민해봐야 할 것 같습니다 

좋은 의견 감사합니다 두루 ☺️

Gyuil-Hwnag replied:
페토의 아주 좋은 의견이네요~👍
사실 이 다음 질문(사용을 함으로써 장점이 뭘까요?)을 위해서 던진것이었는데, 이미 답변이 있군요!
> 규칙과 원칙 또한 특정한 무언가를 더 잘 만들기 위한 일종의 "수단과 방법"인 것 같습니다
그 수단과 방법이 옳다고만은 할 수 없는 것 같아요 정답은 없으니까요 !

Gyuil-Hwnag commented on app/src/main/AndroidManifest.xml:
```diff
@@ -13,14 +16,22 @@
         </activity>
+        <activity
+            android:name=".view.home.seat.SeatActivity"
+            android:exported="false" />
+        <activity
```
exported 는 어떤 특징을 가지고 있을까요?
추가로 `android:parentActivityName="..."` 속성도 있는데, 이건 어떤 기능을 담당할까요~

chanho0908 replied:
## `exported`
+ 다른 앱에 의해서 실행될 때 시작점이 될 수 있는가? 에 대한 설정입니다.
+ true일 경우 모든 앱에서 접근이 가능하지만, flase일 경우 접근이 불가능합니다.

## `android:parentActivityName`
+ 뒤로 가기 버튼을 눌렀을 때 이동할 화면을 지정합니다. 

```kotlin
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.example.myapplication">

    <application
		    ... >
        <activity android:name=".MainActivity2"
            android:parentActivityName=".MainActivity"
            ></activity>
        <activity android:name=".MainActivity">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>

</manifest>
```

예를 들어 다음과 같이 선언해주면 `MainActivty2`에서 뒤로가기 버튼을 누를 경우 `MainActivity`로 이동합니다 !

Gyuil-Hwnag replied:
아주 좋습니다~
이건 최근에 겪었던 이슈가 있었어요.
딥링크등으로 이동을 시키고, 해당 `android:parentActivityName=""`을 재대로 세팅을 안하다보니 딥링크로 Activity 실행시 순서가 보장이 안됐던 이슈가 있어서 한번 개념을 잡고 가면 좋을 것 같아서 남겨두었었습니다😀

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/MovieStore.kt:
```diff
@@ -0,0 +1,95 @@
+import java.time.LocalDate
+
+class MovieStore {
+    val movies: MutableMap<Int, Movie> = mutableMapOf()
```
오호~ 데이터에 대한 세팅을 해주는 곳을 Store 형태로 지어주셨군요!
해당 네이밍이 나오게 됐던 생각들이 궁금합니다ㅎㅎ🤔

chanho0908 replied:
화면에 보여줄 데이터들을 저장하는 **저장소** 라는 의미로 네이밍했습니다 !

Gyuil-Hwnag replied:
그렇군요~ 저장소라는 의미도 좋군요👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/domain/model/ad/Advertisement.kt:
```diff
@@ -0,0 +1,5 @@
+package woowacourse.movie.domain.model.ad
+
+class Advertisement(
+    val imgResource: String = "ad",
+)
```
`val imgResource: String = "ad"` 해당 기능은 어떤 기능을 담당할까요?🤔

chanho0908 replied:
Drawble에 있는 이미지 리소스로 변환하기 위함입니다 !

이와 같이 구현한 이유는, 만약 DataSource가 서버에서 받아오도록 변경 되었을 때,
외부 변경사항에 의한 도메인 수정이 일어나지 않도록 하기 위한 방법을 고민했습니다.

그 결과 서버에서 이미지의 경로를 받아온다면 String -> Url로,
현재와 같이 Drawble Resource를 사용했을 때 다음 메서드를 사용해 적용할 수 있도록 하기 위함입니다 !

```
fun String.toDrawableResourceId(context: Context): Int {
    return context.resources.getIdentifier(this, "drawable", context.packageName)
}
```

Gyuil-Hwnag replied:
아하 그렇군요~
만약 팀의 컨벤션이 정해져서.. drawable 파일의 네이밍들이 전부다 바뀐다면.... 추가로 해당 히스토리에 대해서 깜빡하고 있었다면 어떻게 될까요🫠

chanho0908 replied:
이미지 표시 실패할 것 같네요 ㅠ_ㅠ,,

말씀해주신 사항들도 고려해보고 싶지만 현재 미션에서 도메인을 변경시키지 않는 방법으로
생각해낸 방법은 이 방법이 최선인 것 같습니다

Gyuil-Hwnag replied:
그렇군요~
사실 고민을 해보길 바랬던 방법은 resId : Int 형식으로 로컬 이미지를 받아보는 것이었어요
String 으로 받다보니 String.toDrawable 이라는 한단계 뎁스가 들어가게 되는 내용이 눈에 띄었어요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private fun initView(savedInstanceState: Bundle?) {
+        ViewCompat.setOnApplyWindowInsetsListener(binding.main) { v, insets ->
+            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
+            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
+            insets
+        }
```
해당 기능은 어떤 기능을 담당할까요?🤔
페토의 의견을 남겨주세요~
아마 수업때 나왔던 내용이라서, 적용을 하셨을 것 같은데 이와 관련 해서 최근에 대응을 하느라 고생을 했던 경험이 있어서 코멘트를 남겨두었어요

chanho0908 replied:
### edge-to-edge
+ Android 15 이상을 타겟팅하는 앱은 기본적으로 디스플레이 영역 전체를 활용하며 이를 `edge-to-edge`라고 합니다. 
+ 말 그대로 끝에서 끝 까지 화면 전체 영역을 사용하겠다고 선언하는거에요 !

![image](https://github.com/user-attachments/assets/dcc99ca7-74c5-4236-9e1c-be70af8ee469)

```kotlin
class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
    }
}
```

`enableEdgeToEdge()`는 AndroidX의 `ActivityCompat` 또는 `WindowCompat`에서 제공하는 API로, 시스템 바를 반투명하게 만들고 앱 콘텐츠를 그 뒤로 확장하는 설정입니다.

앱이 SDK 35 이상을 타겟팅하는 경우, 안드로이드 15 이상의 기기에서는 자동으로 `edge-to-edge` 디스플레이가 활성화되어 다음과 같이 앱의 일부가 상태바 뒤에 그려질 수 있습니다.

![image](https://github.com/user-attachments/assets/92a6ee55-4ab1-4c04-8900-b12360f8d258)

이렇게 의도치 않게 앱의 일부가 상태 바 뒤에 그려질 수 있어, `인셋`을 사용해 처리해야 합니다.

### 인셋이란 ?

화면의 특정 영역이 상태바, 네비게이션바와 같은 시스템 UI와 겹치는 위치를 정의하는 값입니다.

안드로이드의 `WindowInsets` API는 다양한 종류의 `Inset`을 제공하여 앱 개발자가 이러한 겹침 문제를 해결할 수 있게 합니다.

```kotlin
ViewCompat.setOnApplyWindowInsetsListener(binding.main) { v, insets ->
    val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
    v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
    insets
}
```

- `setOnApplyWindowInsetsListener`
    - 인자로 전달 받은 레이아웃을 기준으로 `windowInset` 리스너를 설정합니다.
    - 여기서 main은 `XML` 파일 내의 최상위 Layout으로 현재 화면 전체를 기준으로 인셋을 지정합니다.
- `insets.getInsets(WindowInsetsCompat.Type.systemBars())`
    - 현재 화면의 시스템 바와 교차하는 부분의 인셋을 가져옵니다.
    - 왼쪽, 오른쪽, 하단에서 시스템 바가 차지하는 공간을 나타냅니다.
- `setPadding`
    - 계산된 인셋으로 상하좌우 패딩을 지정합니다.

그 결과 앱은 실제로 **상태바 영역까지** 그려지지만 인셋 처리를 통해 **UI 요소가 시스템 바와 겹치지 않도록 패딩을** 줍니다.

chanho0908 replied:
여담으로 프래그먼트 적용 과정에서 Activity와 Fragment에 모두 인셋을 주어 문제가 되었던 적도 있답니다 ㅎㅎ

<img src="https://github.com/user-attachments/assets/c6d355c0-88a4-453e-9337-1bfe8aae751d" width="300" />

Gyuil-Hwnag replied:
오호 엄청나게 정리를 잘해주셨군요~👍
한번 아래의 내용들을 살펴보고, 앞으로 개발을 해나가면서 도움이 됐으면 해서 오버스팩이긴 하지만 최근까지 대응을 하던 내용이라서 기억에 남아있으니, 참고차 내용을 한번 공유 드려볼게요.

### **개요**

OS15이전에는 EdgeToEdge()가 Default = false 였습니다.

물론 EdgeToEdge()를 사용할 수 있도록 Extension가 있기는 했지만요.

그러나, OS15 이상부터는 EdgeToEdge() 가 Default = true가 되었어요.

이걸 대응하기 위한 구현이에요.

### **추가로 발견된 문제들**

사실 이 하나의 구현만으로 모든 내용이 대응이 된다면, imePadding등 여러가지로 고려해볼 내용들이 덕분에 생기게되었어요.

그 외에 StatusBar / NavigationBar Color 등등...

### **임시 방편**

현재로는 강제로 EdgeToEdge세팅을 끌 수 있는 테마의 세팅이 존재하지만, 이건 OS16에서 Deprecated 될 방법이라서, 점차 나아가서는 `ViewCompat.setOnApplyWindowInsetsListener`으로 대응을 하여야 합니다.

그렇게 된다면, 서비스에서 사용하는 모든 화면들을 대상으로 확인을 해봐야 해서, 상당히 공수가 크게 들어가는 작업이에요.

그외에 위에서 언급한 imePadding등까지 고려한다면 사이드 이팩트는 무수히 커질 수 있겠어요.

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -0,0 +1,71 @@
+        initView(savedInstanceState)
+    }
+
+    private fun initView(savedInstanceState: Bundle?) {
+        ViewCompat.setOnApplyWindowInsetsListener(binding.main) { v, insets ->
```
initView 하나의 함수에서 너무 많은 기능을 담당하고 있다는 생각이 들었는데, 페토의 생각은 어떠신가요~

chanho0908 replied:
앗 !!! initView 함수를 항상 다이어트 하는걸 중요하게 생각했는데 이 부분을 놓쳤네요 🥲

완전 뚠뚠보가 맞습니다 ... 다이어트 좀 시키겠습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/StringFormatter.kt:
```diff
@@ -0,0 +1,18 @@
+import java.time.format.DateTimeFormatter
+
+object StringFormatter {
+    fun dotDateFormat(time: LocalDate): String {
+        val formatter = DateTimeFormatter.ofPattern(DATE_TIME_DOT_FORMAT)
+        return time.format(formatter)
+    }
```
참고차 남겨두었습니다~
아래와 같은 형식으로 작성을 하면, `StringFormat.dotDateFormat(time)` 형태의 코드를 아래처럼 변형도 가능합니다!
~~~kotlin
    @JvmStatic
    fun dotDateFormat(time: LocalDate): String {
        val formatter = DateTimeFormatter.ofPattern(DATE_TIME_DOT_FORMAT)
        return time.format(formatter)
    }

// 사용
dotDateFormat(endDate),
~~~

chanho0908 replied:
`@JvmStatic` 어노테이션을 사용하면 자바에서 다음과 같은 차이가 발생하는군요 !
```
// before
StringFormatter.INSTANCE.dotDateFormat(time);

// after
StringFormatter.dotDateFormat(time);
```

이전 미션에서 메서드를 사용하는 곳에서 StringFommater를 호출하지 않고 사용하다
정적 import는 이 함수가 확장 함수인지, top-level인지, 정적 메서드인지 알 수 없다는 문제가
발생한다는 리뷰를 받고 이 의견에 동의해 기존 구현을 유지하겠습니다 ! ☺️

좋은 의견 감사합니다 두루 !

Gyuil-Hwnag replied:
오호~ 좋습니다👍
페토의 의견이 담긴 내용 정말 좋습니다

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/ext/IntentExt.kt:
```diff
@@ -0,0 +1,16 @@
+import android.os.Build
+import java.io.Serializable
+
+inline fun <reified T : Serializable> Intent.getSerializable(
+    key: String,
+    clazz: Class<T>,
+): T {
+    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        getSerializableExtra(key, clazz) ?: throw IllegalArgumentException()
+    } else {
+        getSerializableExtra(key) as T
+    }
+}
```
Serializable의 Extension 내용 너무 잘해주셨네요👍
단순 `IllegalArgumentException()`를 하는 것이 아닌, 에러에 대한 메시지를 남겨둔다면 실제 규모가 큰 프로젝트에서 크래시가 발생했을 때 어떤 부분에서 발생을 했는지 좀 더 유추하기 쉬울 수 있겠다는 생각도 들었어요~

chanho0908 replied:
에러 메시지까지 출력하도록 수정해보겠습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/ext/StringExt.kt:
```diff
@@ -0,0 +1,7 @@
+
+import android.content.Context
+
+fun String.toDrawableResourceId(context: Context): Int {
+    return context.resources.getIdentifier(this, "drawable", context.packageName)
+}
```
해당 기능이 들어간 이유가 궁금합니다~🤔

chanho0908 replied:
https://github.com/woowacourse/android-movie-theater/pull/102#discussion_r2070467224

해당 코멘트에 답변했습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/ext/bindingadapter/BindingAdapterExt.kt:
```diff
@@ -0,0 +1,10 @@
+@BindingAdapter("imageFromName")
+fun ImageView.setImageFromName(name: String) {
+    setImageResource(name.toDrawableResourceId(context))
+}
```
name으로 구분을 한다는 것이 뭔가 눈에 들어오긴 했어요~
실제 프로젝트라고 생각을 해본다면, url / localRes를 사용하도록 할텐데 이에 맞게 변경을 해보려면 어떻게 해볼 수 있을까요?🤔

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/HomeFragment.kt:
```diff
@@ -0,0 +1,84 @@
+    override fun onDestroyView() {
+        super.onDestroyView()
+        _binding = null
+    }
```
onDestroyView시에 _binding = null 세팅 정말 좋습니다👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/booking/AdapterItemSelectedListener.kt:
```diff
@@ -0,0 +1,23 @@
+import android.widget.AdapterView
+
+class AdapterItemSelectedListener(
+    private val action: (Int) -> Unit,
```
해당 객체는 한번 클릭한 경우 다시는 클릭을 못하도록 하는 기능인가 보군요~
프로젝트가 커진다면, 요구사항에 의해서 Utils등으로 여러곳에서 사용을 해볼 수 있을 것 같은데 의미있는 네이밍을 지어볼 수 있겠다는 생각이 들었어요!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/booking/BookingPresenter.kt:
```diff
@@ -0,0 +1,100 @@
+
+        if (availableTimes.isEmpty()) {
+            view.showToast()
+            return
```
이 부분은 코드레빗의 내용도 좋은 것 같아서 한번 반영을 해보면 좋겠다는 생각이 들었어요~
[코드레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/102/files#r2069927814)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/movies/viewholder/TheaterViewHolder.kt:
```diff
@@ -0,0 +1,24 @@
+        count: Int,
+    ) {
+        binding.model = item
+        binding.root.setOnClickListener {
+            onClick(item)
+        }
```
`onClickItem`도 DataBinding을 사용한다면, 구현을 줄여볼 수 있어서 한번 고민을 해보는 것도 좋을 것 같다는 생각이 듭니다!
- 예시 입니다! (예시이므로, 적용하다 어려움이 있으면 언제든 말씀 주세요!)
~~~kotlin
interface ActionHandler {
   fun onNextButtonClicked()
   fun onPreviousButtonClicked()
}

// xml
<data>
   <variable
      name="eventListener"
      type="ActionHandler">
   </variable>
</data>

<View
   android:onClick="@{() -> eventListener.onNextButtonClicked()}" />
<View
   android:onClick="@{() -> eventListener.onPreviousButtonClicked()}" />

// Activity
class Activity: CartActionHandler {
   override fun onNextButtonClicked() {}
   override fun onPreviousButtonClicked() {}
}

// Activity or Adapter kt
class Adapter (actionHandler: ActionHandler)
// ViewHolder
binding.eventListener = actionHandler

~~~

chanho0908 replied:
한번 공부해보고 적용해보겠습니다 :)

Gyuil-Hwnag commented on app/src/main/res/layout/activity_booking.xml:
```diff
@@ -0,0 +1,169 @@
+            android:layout_height="wrap_content"
+            android:layout_marginHorizontal="12dp"
+            android:gravity="center"
+            android:textSize="20sp"
+            android:textStyle="bold"
```
sp와 dp는 어떤 특징들을 가지고 있을까요~🤔

chanho0908 replied:
dp는 기기 폰트 크기 설정에도 고정된 크기를, sp는 사용자 시스템 설정에 따라 폰트 크기가 변경됩니다 !

일전에 프로젝트를 하면서 앱의 도메인이 무장애 여행지 어플이라 접근성이 매우 중요했었습니다.

그래서 애초에 앱을 접근성 위주로 설계하다보니 이미지나 버튼같은게 커져서 기기 설정에 폰트 크기가 영향을 받으면
화면이 너무 이상해져 sp 대신 dp 사용한 아에 글자 크기를 고정한 기억이 나네요 🤔

Gyuil-Hwnag replied:
오호 페토의 경험이 담긴 내용 정말 좋네요👍

실제로 개발을 하면서 은근히 고민이 많이 되는 부분들이라서 한번 알아두고 간다면, 디자이너분과 얘기를 하면서 좀 더 좋은 사용성 & 디자인을 제공할 수 있을 것 같습니다ㅎㅎ

Gyuil-Hwnag commented on app/src/main/res/layout/activity_seat.xml:
```diff
@@ -0,0 +1,207 @@
+
+                <TextView
+                    android:id="@+id/a4"
+                    style="@style/SeatPurple"
+                    android:text="A4" />
```
style을 정의해주셨군요!!👍
아주 좋네요ㅎㅎ
해당 내용을 통해서 공통된 내용들도 줄일 수 있고, 공통된 디자인을 제공할 수 있겠군요~

Gyuil-Hwnag reviewed (changes requested):
페도 안녕하세요~
레벨1에 이어서 다시 만나게 됐군요 잘부탁드립니다!

해당 미션 구현 고생 많으셨습니다!
미션 구현을 너무 잘 해주셨습니다 💯

이번 미션을 진행하면서, DataBinding 부분이 어려움이 있으셨군요! 해당 내용은 저랑 이번 미션을 진행하면서 같이 얘기를 해보도록 해봅시다~💪
추가로 고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
이번 미션도 화이팅입니다💪

Gyuil-Hwnag commented:
페토 안녕하세요~
남겨주신 질문이 있어서 내용 남겨두었어요!

> 첫 번째로, XML에 로직이 들어가는게 굉장히 어색하게 느껴졌어요 🧐
아마 데이터 바인딩을 사용하는 이유 중 하나로 Activity나 다른 뷰에 존재해야할 코드들이 줄어든다는 장점이 있을 수도 있지만
결국 이 코드들도 XML에 존재하고 이렇게 코드가 분산되는게 저는 영 맘에 들지 않았답니다.

어떤 의미인지 너무 와닿는군요~
개인적으로 DataBinding을 잘못 활용을 하게 된다면, 오히려 하나의 구현을 볼때 Activity, xml 등 너무 많은 파일들을 왔다갔다 해야 해서, 더 구현을 알아보기 힘들다는 단점이 있는 것 같아요!

그렇다면, 과연 장점을 잘 활용하기 위한 방법들은 어떤 것들이 있을까요?
개인적으로는 첫 출발은 `공통으로 사용할 수 있는 DataBinding 을 만들어 본다!` 일 것 같아요
예시로는 Image, Int to String등이 있을 것 같다는 생각이 들었어요~
이번 단계를 통해서 DataBinding에 대해서 알아보고, 다음미션들을 통해서 DataBinding과 엮을 수 있는 다른 것들을 본다면 DataBinding의 장점들이 더욱더 와 닿을 수 있으니, 차근차근 한단계식 밟아가 나아가는 것도 좋을 것 같다는 생각이 들었습니다😀

> 두 번째로, BindingAdapter 같은 불필요한 코드들이 너무 많아진다고 생각해요
그냥 setImageResource 같이 기존에 제공하는 메서드들을 사용하면 되는데 이걸 꼭 만들어야할까 ?..

현재는 로컬의 이미지를 단순히 보여주는 것이라서 간단할 수 있지만, 실제로 프로젝트를 하다보면 하나의 이미지를 보여줄 때도 캐싱 처리 / 오류시 화면등 여러가지로 처리해야할 게 늘어날 수 있어요!
이러한 것들을 적용을 해보지는 않지만, 위에서 얘기한대로 `공통으로 사용할 수 있는 DataBinding 을 만들어 본다!`의 느낌으로 해보는 것도 좋을 것 같다는 생각이 들었어요~
방법은 여러가지가 있을 것 같아요
DataBinding 에는 파라미터를 하나만 받을 수 있는 것이 아니라서, Local, Url 을 한번에 처리할 수 있는 로직을 만든다! 등도 해볼 수 있을 것 같았어요

chanho0908 committed:
refactor : 영화 정보 초기화 로직 Prestenter의 init block으로 이동

- 홈 프래그먼트 패키지 수정

chanho0908 committed:
refactor : 영화 목록 (홈) 프래그먼트 네이밍 수정

chanho0908 committed:
feat : Toast Message 출력 확장함수 추가

chanho0908 committed:
refactor : intent 전달시 빈 값일 경우 에러 메시지를 출력하고 화면을 종료하도록 수정

chanho0908 committed:
refactor : 단순 코드 개선

- 영화 저장소의 영화 상태를 저장하는 map의 가시성 변경
- 극장 저장소의 상영 날짜 더미 데이터 계산 로직에서 마지막 날이 중복되어 들어가는 현상 수정
- 좌석 View 객체에서 전달받은 Seat를 불필요하게 다시 Seat로 매핑하는 코드 제거

chanho0908 committed:
refactor : 리사이클러뷰용 아이템 모델 네이밍 수정

- Domain Model -> Item 매핑 함수 패키지 분리

chanho0908 committed:
refactor : 영화 & 광고 리사이클러뷰 코드 개선

chanho0908 committed:
feat : 특정 극장의 특정 영화 상영 정보를 탐색 기능 추가

- 상영 일정 Fixture 추가
- 영화관 Fixture 추가

chanho0908 committed:
feat : ImageBindingAdapter의 여러가지 이미지 소스 대한 처리 기능 추가

chanho0908 committed:
feat : 영화 목록 리사이클러뷰 클릭 리스너 데이터 바인딩 적용

chanho0908 committed:
feat : 극장 목록 리사이클러뷰 데이터바인딩 적용

chanho0908 committed:
refactor : 영화 객체 스스로의 모든 상영 시간 상태 프로퍼티 제거, 상영 시작일, 종료일만 가지도록 수정

chanho0908 committed:
feat : 영화관별로 다른 상영시간을 갖도록 수정

chanho0908 committed:
feat : 영화 목록 presenter 팩토리 메서드 추가

chanho0908 committed:
feat : 예매 화면 presenter 팩토리 메서드 추가

chanho0908 committed:
feat : 영화 예매 화면의 영화 기본 정보 데이터 바인딩 적용

chanho0908 committed:
feat : 예매 화면 인원 증감 버튼 데이터 바인딩 적용

chanho0908 committed:
refacor : 영화 정보 초기화시 예매 가능한 시간 정보도 함께 presenter에서 초기화하도록 수정

chanho0908 committed:
refactor : 예약 정보 객체의 영화 제목 프로퍼티명 수정

chanho0908 committed:
refactor : 상영 날짜 객체 코드 개선

chanho0908 committed:
feat : 예매 인원 객체에 프로퍼티 변경 함수 추가

chanho0908 committed:
feat : 기본 영화 상영 정보 객체의 상영 날짜 추출 함수 추가

chanho0908 committed:
feat : 예매 화면 데이터 바인딩 적용 & 예약 정보를 날짜와 시간을 선택할 때 마다 저장하도록 수정

chanho0908 committed:
refactor : 영화 목록 화면 presenter 주생성자 private 해제

chanho0908 committed:
refactor : 예매 화면에서 영화 러닝 타임에 "분" 이 출력되지 않는 문제 수정

chanho0908 committed:
refactor : 극장 목록의 "n개의 상영 시간" 포맷팅을 xml에서 하도록 수정

chanho0908 committed:
feat : 영화관의 n 개의 상영 시간을 출력할 때 모든 상영 시간개수가 아닌 유니크한 시간 개수로 수정

chanho0908 committed:
refactor : 특정 영화관에서 상영하지 않는 영화가 존재하도록 수정

chanho0908 committed:
refactor : Intent Serializable 확장함수 수정

chanho0908 committed:
feat : kotlin-parcelize dependency add & Bundle Parcelable 확장함수 추가

chanho0908 committed:
feat : 영화관 바텀시트 출력 후 화면 전환시 앱 크래시 되는 현상 수정

chanho0908 committed:
feat : ArrayList Serializable을 받기 위한 Bundle 확장함수 추가

chanho0908 committed:
feat : 좌석 객체에 선택된 좌석 반환 함수 추가

chanho0908 committed:
feat : 좌석 선택중 화면 회전시 복구 기능 추가

chanho0908 committed:
chore : 기타 코드 수정

chanho0908 committed:
refactor : 메인 화면 initview 함수 분리

chanho0908 committed:
feat : 좌석 선택 화면 회전시 가격 복구기능 추가

chanho0908 commented on app/src/main/java/woowacourse/movie/view/home/movies/MovieListPresenter.kt:
```diff
@@ -47,6 +47,12 @@ class MovieListPresenter(
+        fun initialize(view: MovieListContract.View): MovieListContract.Presenter {
+            val movieStore = MovieStore()
+            val theaterStore = TheaterStore(movieStore)
+            return MovieListPresenter(view, movieStore, theaterStore)
+        }
```
View가 dataSource까지 알게되는 것은 과하다고 생각해서 
Store들만 팩토리 메서들을 통해 주입하도록 했습니다 !

Gyuil-Hwnag replied:
오호~ Factory 형태로 생성을 해주셨군요!
이 방법도 좋네요👍
추후 Store가 변경됐을 때 아주 유용하게 사용을 해볼 수 있겠어요

다만, 한가지 한번 질문을 던져 볼게요~
`전역 변수로 선언 vs 생성자로 넣고 팩토리 함수로 생성` 둘중, 2번으로 선택을 하신 이유가 궁금합니다🤔

chanho0908 replied:
저는 싱글톤이나 top-level 함수를 가급적 지양하는 편입니다 ! 
(앞서 말씀 드린 것 처럼 확장 함수는 일반적 top-level 함수와는 다르다고 생각해요 !)

이유는 개발하는 과정에선 어디서든 꺼내 쓸 수 있기 때문에 편리하지만
반대로 객체지향적으로 생각했을 때 경계가 없는, 캡슐화 되지 않는 객체라고 생각해서 사용하지 않았습니다 !

Gyuil-Hwnag replied:
그렇군요 페토의 의견 아주 좋습니다~
뭔가 객체지향적으로 만들려다 보니, 더욱 구현이 복잡해진 느낌이 살짝 드는건 어쩔 수 없나 보군요

chanho0908 commented on app/src/main/java/woowacourse/movie/view/home/movies/MovieListFragment.kt:
```diff
@@ -41,8 +43,13 @@ class MovieListFragment : Fragment(R.layout.fragment_home), MovieListContract.Vi
-        theaterAdapterEventHandler = TheaterAdapterEventHandler(presenter)
         presenter.loadUiData()
+
+        setFragmentResultListener(KEY_REQUEST) { _, bundle ->
+            bundle.getParcelableOrNull<Result>(KEY_RESULT)?.let {
+                presenter.loadMovieScreening(it.movieId, it.theaterName)
+            }
+        }
```
기존에 바텀시트 프래그먼트의 주 생성자에 파라미터를 주입받는 방식으로 사용했는데,
화면 전환 시 앱이 크래시되는 문제를 발견했습니다!

이를 조사해본 결과, 화면 전환 시 액티비티는 파괴되고
onSaveInstanceState() 내부에서 FragmentManager가 프래그먼트 상태를 저장합니다.

```kotlin
protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    Parcelable p = mFragments.saveAllState();
    if (p != null) {
        outState.putParcelable(FRAGMENTS_TAG, p);
    }
}
```
이후 복원 단계에서 FragmentManager는 리플렉션을 사용해 프래그먼트 인스턴스를 생성하는데,
이때 반드시 매개변수 없는(public no-arg) 생성자가 필요합니다.

하지만 기존에 주 생성자에 파라미터가 있었기 때문에,
해당 인자들이 복원되지 못하고 InstantiationException이 발생하면서 앱이 크래시 되는 문제가 발생합니다 !

Gyuil-Hwnag replied:
오호~ 크래시 발견 -> 원인 파악 -> 수정까지 정말 완벽합니다💯

chanho0908 commented on app/src/main/java/woowacourse/movie/view/base/BaseViewHolder.kt:
```diff
@@ -0,0 +1,15 @@
+package woowacourse.movie.view.base
```
지난주 팝업 세션에서 디노께서 리사이클러뷰를 얼마나 잘 만들 수 있는지
공유해주셨는데요 !

제가 이해하고 적용할 수 있는 부분까지만 제 방식대로 해석해서 수정해봤습니다 !

chanho0908 reviewed (commented):
안녕하세요 두루 !

이번에 정말 많은 변경사항이 있었던 것 같습니다 🙃

특히 그 동안 데이터 바인딩을 적용하지 못했던 부분들을
두루께서 알려주신 내용을 바탕으로 꼼꼼히 적용해봤습니다

데이터 바인딩이 너무 어렵게 느껴졌는데 두루 덕분에 많이 친해질 수 있었어요 ! ☺️

코드레빗이 푸시하면 바로 리뷰 달아주는걸 모르고 푸시하다가
여기저기 코드레빗 리뷰가 달려서 다소 지저분해진 것 같습니다... 🥲🥲

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -60,4 +68,6 @@ dependencies {
     androidTestImplementation(libs.kotest.runner.junit5)
     androidTestImplementation(libs.mannodermaus.junit5.core)
     androidTestRuntimeOnly(libs.mannodermaus.junit5.runner)
+    implementation(libs.androidx.recyclerview)
+    implementation(libs.androidx.fragment.ktx)
```
아마 프로젝트를 설정할 때부터 version catalogs를 사용하도록 되어 있었을 것 같은데, 이전에는 gradle을 kotlin 이전에 groovy로 적던 시절도 있었어요~
오래된 프로젝트일 수록 마이그레이션이 이뤄지지 않은 경우 groovy, 또는 version catalogs 가 적용이 안된 프로젝트일 때도 있답니다!
이전에는 어떻게 gradle 을 정의했었는지도 추후에 시간이 남을 때 살펴본다면 많은 도움이 될 것 같다는 생각이 듭니다💪

지금 당장 하기에는 너무 과한 내용일 것 같아서 추후에 한번 시간이 남을 때 살펴보시면 좋을 것 같아요

- groovy -> groovy to kotlin -> buildSrc등 version catalogs 이전 gradle 관리법 -> version catalogs -> custom plugins

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -0,0 +1,78 @@
+package woowacourse.movie.view
+
```
만약 서비스가 커진다면, 페토는 어떻게 패키지를 나누실지 궁금합니다~

개인플젝으로 했었던 내용을 한번 예시로 드려볼게요!
(아래의 구조를 보고 어떤 생각이 드시나요! 이렇게 하면 너무 세분화 돼서 보기 힘들어서 구리다!!! 이런 피드백도 좋아요
참고로 core: base 및 utils등, ViewModel : Presenter로 생각해보시면 될 것 같아요)
<img width="335" alt="스크린샷 2025-05-05 오전 12 34 09" src="https://github.com/user-attachments/assets/08b40034-8658-40fe-81f0-2092b81a8375" />

<img width="274" alt="스크린샷 2025-05-05 오전 12 37 50" src="https://github.com/user-attachments/assets/e6385d5b-d732-4bb9-b37c-adedce89a3dd" />

chanho0908 replied:
오랜만에 멀티 모듈을 보니 가슴이 뛰네요 😍

우테코에 들어오기전에 한 달 정도 시간이 비어서 멀티 모듈을 혼자 공부한 적이 있는데
정말 정말 정말 어려웠던 것 같아요 😭

저도 패키징은 매번 고민이 되는 문제인 것 같아요 할 때 마다 생각이 바뀌는 것 같은...

제 경험에 기반해 말씀드리면,

저는 우선 동일한 플로우를 타는 화면은 하나의 패키지에 묶는게 좋다고 생각합니다.
영화 예매를 예로 들면,
  - booking
    - movies
       - MovieListFragment.kt
       - MovieListPresenter.kt
       - MovieListContract.kt
 - setting
    -  myInfo
    - ...
이런식으로 구현을 할 것 같습니다

Gyuil-Hwnag replied:
사실 이 코멘트는 멀티 모듈 보다는 모듈 == 패키지로 생각해서 나눈 내용을 살펴보시면 좋을 것 같습니다~
(이 설명이 빠져있었군요😅)

> 저는 우선 동일한 플로우를 타는 화면은 하나의 패키지에 묶는게 좋다고 생각합니다.

오호 그렇군요! 그것도 아주 좋은 의견 이네요ㅎㅎ
다만 서비스의 플로우가 변경된다면, 최신화를 해줘야 겠군요!

그렇다면, 여기서 나온 의견대로 패키지 구조를 바꾼다면 어떻게 바뀌여야 할까요??
<img width="332" alt="스크린샷 2025-05-05 오후 9 50 45" src="https://github.com/user-attachments/assets/acf58c74-313c-49ab-b69b-1b8bcbefa7f2" />

페토의 자유로운 의견을 남겨주세요ㅎㅎ
혹시나 도움이 될까 참고차 저라면 어떻게 해봤을까 참고차 의견을 남겨볼게요!

- view
    - core
        - base
        - ext
        - bindingadapters
    - home
        - model (optional)
            - uiModels / mapper
        - HomeActivity
        - HomeContract
        - HomeBindingAdapters
        - HomeAdapter
        - HomeActionHandler
...

chanho0908 replied:
패키징 수정하면서 나눈 커밋들에 의견들을 첨부하겠습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -0,0 +1,78 @@
+import woowacourse.movie.view.home.movies.MovieListFragment
+import woowacourse.movie.view.setting.SettingFragment
+
+class MainActivity : AppCompatActivity() {
+    private lateinit var binding: ActivityMainBinding
```
MainActivity 라는 네이밍이 살짝 눈에 들어왔군요~
만약 서비스가 커져서, 서비스 메인이 따로 생긴다면 해당 Activity의 네이밍은 어떻게 바뀔 수 있을까요?

chanho0908 replied:
제가 생각하기에 MainActivity는 유지를 하는게 맞을 것 같아요

MainActivity라는 이름이 명시적으로 봤을 때 앱의 시작점(스플래시, 로그인 등등...) 을 지나고
딱 화면에 나오는 첫 화면이라는 인식이 강해서 저는 유지하는 것이 맞을 것 같습니다 !

Gyuil-Hwnag replied:
페토의 의견이 그렇다면 아주 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/base/BaseViewHolder.kt:
```diff
@@ -0,0 +1,15 @@
+import androidx.annotation.LayoutRes
+import androidx.recyclerview.widget.RecyclerView
+
+abstract class BaseViewHolder<T>(
+    parent: ViewGroup,
+    @LayoutRes layoutRes: Int,
+) : RecyclerView.ViewHolder(
+        LayoutInflater.from(parent.context).inflate(layoutRes, parent, false),
+    ) {
+    abstract fun bind(item: T)
+}
```
BaseViewHolder를 적용해주셨군요~
개인적으로 Base는 지양을 하긴 하는데, 이유는 개발을 하다가 이것저것 추가하다보면 Base에 너무 많은 의존을 가지게 될 수 있어서입니다.
이렇게 되지 않기 위해서는 Base에는 정말 기본적인 것을 추가하면 추가하는 것들에 대해서 페토만의 기준을 확실히 세우는 것이 도움이 되겠군요💪

chanho0908 replied:
두루의 의견에 오만퍼센트 동의합니다 😁

일전에 동료 크루분께서 BaseActivity를 사용하셨고 이 것에 대한 토론을 한 적이있는데 대게 이 코드가 들어가 있더군요 !
```
ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
    val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
    v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
    insets
}
```
저는 BaseXXX 를 사용한다면, 정말 이 Base에 의존하는 곳이 모든 곳에서 동일한 기능을 하는가?
라고 고민해봤을 때 저는 이 경우가 드물것이라고 생각했습니다.

앞서 두루가 말씀해주신 것 처럼 화면에 따라서 인셋을 다르게 적용한다면 Base에 의존하는 구조는
자연스래 깨질 수 밖에 없다고 생각해요

그렇게 되면 어디선 Base를 사용하고 어디선 사용하지 않는 구조가 되며 이는 일관성이 깨진다고 생각해
저 또한 Base를 지양합니다(~~BaseActivity, BaseFragment를 보면 눈이 돌아갑니다😡~~)

그래서`ViewHoder` 또한 정말 최소한의 inflate를 위한 코드만 존재하도록 했는데,
사실 bind 함수에 과연 모든 바인딩 로직을 담을 수 있을까? 에 대한 의문이 남긴합니다  🤔

Gyuil-Hwnag replied:
오호~ 페토의 경험이 담긴 내용 아주 정말 좋네요👍
> 사실 bind 함수에 과연 모든 바인딩 로직을 담을 수 있을까? 에 대한 의문이 남긴합니다 🤔

결국 그러다 보니... 여러가지 Base등이 계속 생겼던 일이 있었네요..🥲
(어떤 View는 클릭도 있고, 만약 상태를 변경하는 로직들도 있다보니... Base가 점점 생겨나는 마법이 발생하더라구요)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/bindingadapter/ImageBindingAdapter.kt:
```diff
@@ -0,0 +1,19 @@
+        is ImageSource.Url -> {}
+        is ImageSource.Resource -> setImageResource(source.name.toDrawableResourceId(context))
+    }
+}
```
sealed class로 추가를 해주셨군요~
이전에 남겼던 코멘트에는 아래와 같은 내용으로 BindingAdapter에는 여러개의 파라미터를 넘길 수 있다는 것을 살펴보게 위함이었어요💪
~~~kotlin
@BindingAdapter(value = ["imageUrl", "imageRes"], requireAll = false)
fun ImageView.setImage(imageUrl: String?, imageRes: Int?) {
    if (imageUrl.isNullOrBlank().not()) { } else if (imageRes != null) { }
}
~~~

chanho0908 replied:
이렇게도 사용할 수 있군요 !

두루덕에 새로운걸 알아갑니다 감사합니다 ☺️

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/ext/BundleExt.kt:
```diff
@@ -0,0 +1,36 @@
+
+@Suppress("UNCHECKED_CAST")
+inline fun <reified T : Serializable> Bundle.getSerializableArrayList(key: String): ArrayList<T> {
+    return (
```
반드시 해당 내용에서 ArrayList 여야 하는 이유가 있을까요~
추가로 [코드레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/102/files#r2072601240) 해당 내용도 한번쯤은 살펴봐도 좋을 것 같다는 생각이 들었습니다!

chanho0908 replied:
이 부분은 아직 여러 자료를 찾아보았지만 명확하게 이해하기 조금 어려웠던 것 같습니다 😓

우선 `List`는 인터페이스 `ArrayList`는 구현체라는 점이 가장 큰 차이인 것 같습니다 

직렬화는 객체를 저장 가능한 형태(예: 바이트 스트림)로 바꾸는 작업인데, 인터페이스는 그 자체로 직렬화 대상이 될 수 없습니다.

또한 `ArrayList`는 내부적으로 Serializable을 구현하고 있기 때문에 ArrayList를 사용해야 한다고 이해했습니다 ! 🙃
![image](https://github.com/user-attachments/assets/c443259b-4781-4bfe-80f5-5ecef5bb03e4)

Gyuil-Hwnag replied:
오 아주 좋네요ㅎㅎ👍
내부적으로 Serializable 이 구현된 경우에 사용이 가능합니다
(한번 추후에 Serializable이 가능한 것과 불가능한 것에 대해서 정리를 해보면 좀 더 유용하게 사용을 해볼 수 있겠다는 생각이 들었어요ㅎㅎ)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/ext/IntentExt.kt:
```diff
@@ -0,0 +1,13 @@
+import android.os.Build
+import java.io.Serializable
+
+inline fun <reified T : Serializable> Intent.getSerializableOrNull(key: String): T? {
+    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        getSerializableExtra(key, T::class.java)
+    } else {
+        getSerializableExtra(key) as? T
+    }
+}
```
`non-nullable (require) / nullable (optional)`로 구분을 한다면, 필수로 넘겨야 하는 내용들에 대해서 null-check 가 중복적으로 들어가는 부분을 줄여볼 수 있겠다는 생각이 들었어요~💪

chanho0908 replied:
좋은 의견 감사합니다 두루 !

getSerializableExtra 메서드가 내부적으로 null을 반환할 수 있기 때문에 null을 반환하지 않기 위해선 반드시 에러처리가 필요합니다.

<img width="696" alt="image" src="https://github.com/user-attachments/assets/09d681b0-a204-4d98-8959-ba28f631195b" />

```
package woowacourse.movie.view.ext

import android.content.Intent
import android.os.Build
import java.io.Serializable

inline fun <reified T : Serializable> Intent.requireSerializable(key: String): T {
    val value = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
        getSerializableExtra(key, T::class.java)
    } else {
        @Suppress("DEPRECATION")
        getSerializableExtra(key) as? T
    }
    return value ?: error("Serializable extra for key \"$key\" is missing in Intent.")
}
```
이와 더불어 kotlin의 null을 반환하는 api들에 대한 오랜 고민이 있는데요,

논리적으로 예외가 발생하지 않는 상황이라면 그냥 `throw IllegalArgumentException` 같이 에러를 반환하도록
구현하는게 맞을까요 ?

예를 들어, 
+ 다른 액티비티로부터 직렬화한 데이터를 받을 때 저는 데이터가 넘어오지 않는 예외가 과연 발생할까?
+ `Map`의 `key`로 `value`를 찾을 때 `key`에 반드시 해당하는 값만 들어간다면 에러가 발생할까 ?

이런 경우에 저는 보통 에러가 절대 발생하지 않는 상황일거라고 생각해서 `throw IllegalArgumentException` 을 사용해서
null 처리를 하는데요 ! 

논리적으로 예외가 절대 발생하지 않는 상황이라도 에러가 발생했을 때를 대비한 구현을 하는게 맞을지
에러를 던지도록 처리할지 고민이 됩니다 

두루가 생각하시기에 이런 상황에선 어떻게 구현하는것이 좋은 방법일까요 ?

Gyuil-Hwnag replied:
사실 `nullable하게 반환 + non-null 일 경우 처리` 이렇게 있으면, 크래시가 발생해도 대응은 가능합니다!
다만, 제가 위와 같이 `non-nullable (require) / nullable (optional)` 남겨보도록 했던 이유는 아래와 같았어요
1. 추가적으로 필수로 넘겨야 하는 값들에 대해서 모든 Serializable로 받는 곳에서 null-check 를 하는 보일러 플레이트 구현을 줄여보도록 하기 위함이었어요.
2. 필수 값인 경우에 값을 안넘겨서 에러가 발생한다면, 이건 개발자가 로직상 잘못 짰던 이유가 큽니다. 
- 발생하면 안되는 오류이기 때문이지요.. 만약 발생을 한다면, 에러 메시지를 던지게 하고 그걸 로그로 남겨둔다면 어떤 부분에서 크래시가 발생하는지를 좀 더 잘 캐치 하기 위함 이었어요. 
- null check 하는 부분에서도 물론 해도 괜찮긴 하지만요.

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/booking/BookingContract.kt:
```diff
@@ -0,0 +1,49 @@
+import java.time.LocalTime
+
+interface BookingContract {
+    interface View {
```
Contract의 생성및 MVP 에서의 View / Presenter의 역할에 대한 구분 너무 잘해주셨네요!👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/complete/BookingCompleteActivity.kt:
```diff
@@ -0,0 +1,124 @@
+
+class BookingCompleteActivity : AppCompatActivity(), BookingCompleteContract.View {
+    private lateinit var binding: ActivityBookingCompleteBinding
+    private lateinit var presenter: BookingCompleteContract.Presenter
```
BookingActivity 처럼 해당 부분도 DataBinding을 활용하도록 해볼 수 있을것 같다는 생각이 드는데, 페토의 의견은 어떠신가요~💪

chanho0908 replied:
앗 좌석 선택 화면과 예약 완료 화면에 데이터 바인딩 적용하는 것을 깜빡했네요😅  적용하겠습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/movies/model/MovieRvItem.kt:
```diff
@@ -0,0 +1,27 @@
+        val imgResource: ImageSource,
+    ) : MovieRvItem(ViewType.VIEW_TYPE_ADVERTISEMENT)
+
+    enum class ViewType(
+        @LayoutRes val layoutRes: Int,
+    ) {
+        VIEW_TYPE_ADVERTISEMENT(R.layout.advertisement_item),
+        VIEW_TYPE_MOVIE(R.layout.movie_item),
+    }
```
ViewType으로 만들면서 가져가는 이점은 어떤 것이 있을까요~🤔
페토의 의견이 궁금합니다!

개인적으로는 아래의 생각들이 고민해 볼 포인트로 생각났어요
1. 만약 타입이 늘어난다면, `Adapter 내의 코드 추가 & ViewType 추가`
2. 해당 레이아웃을 찾으려면 Adapter -> Type -> layout 으로 뎁스의 증가

chanho0908 replied:
`ViewType enum class`를 만들지 않으면 다음과 같은 문제가 발생할 것 같습니다

- onCreateViewHolder의 else 
- 0과 1의 매직 넘버(상수화 할 순 있겠지만 else는 동일)

```
override fun getItemViewType(position: Int): Int {
    return when (movieRvItems[position]) {
        is MovieRvItem.MovieItem -> 0
        is MovieRvItem.AdItem -> 1
    }
}

override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<MovieRvItem> {
    return when (viewType) {
        0 -> AdvertiseViewHolder(parent, R.layout.advertisement_item)
        1 -> MovieViewHolder(parent, R.layout.movie_item, handler)
        else -> throw IllegalArgumentException("Invalid viewType: $viewType")
    }
}
```

반면 `ViewType enum class`를 만든다면 `ordinal`은 `ViewType` 기준으로 고정되어 있어서
잘못된 숫자 비교하는 문제를 피할 수 있고 매직 넘버도 제거할 수 있을 것 같습니다 

다만, 디노의 코드를 참고하면서 다르게 적용한 부분이 있습니다.

원본 코드에선 `enum class`에 `layout id`를 넣지 않았는데요,  `layout id`를 넣게된다면
두루가 말씀하신 두 가지 문제가 모두 발생합니다

제가 생각한 `layout id`을 넣었을 때의 이점은 다음과 같습니다.
+ Adapter가 구체적인 layout Id를 직접적으로 모를 수 있다.
+ oncreateViewHolder에서 ViewType을 분기하고 바로 사용할 수 있다
```
override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): BaseViewHolder<MovieRvItem> {
        return when (ViewType.entries[viewType]) {
            ViewType.VIEW_TYPE_ADVERTISEMENT -> {
                AdvertiseViewHolder(
                    parent,
                    R.layout.advertisement_item
                )
            }
            ViewType.VIEW_TYPE_MOVIE -> {
                MovieViewHolder(
                    parent,
                    R.layout.movie_item,
                    handler
                )
            }
        }
    }
    
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): BaseViewHolder<MovieRvItem> {
        return when (val type = MovieRvItem.ViewType.entries[viewType]) {
            MovieRvItem.ViewType.VIEW_TYPE_ADVERTISEMENT ->
                AdvertiseViewHolder(
                    parent,
                    type.layoutRes,
                )

            MovieRvItem.ViewType.VIEW_TYPE_MOVIE ->
                MovieViewHolder(
                    parent,
                    type.layoutRes,
                    handler,
                )
        } as BaseViewHolder<MovieRvItem>
    }

```
저는 위 두 가지 이유로 ViewType enum class가 layout id를 갖게 했는데요 ,

두루가 말씀해주신 내용들도 충분히 고민해볼 포인트라 아마 페어나 팀원들의 의견에 따라 사용할 것 같습니다 ☺️

Gyuil-Hwnag replied:
오! 이미 많은 고민을 거쳐서 나온 내용이었군요~
해당 객체는 Android에 대한 의존성이 점점 커진다는 점도 있긴 하지만, 뭔가 너무 딥하게 갈 것 같고 페토도 많은 고민을 해서 나왔던 내용인 것 같으니, 넘어가보도록 해볼게요!👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/seat/SeatPresenter.kt:
```diff
@@ -0,0 +1,60 @@
+        view.moveToBookingComplete(ticket)
+    }
+
+    override fun restore(seat: ArrayList<Seat>) {
+        this.seats.restore(seat)
+        view.showSeat(seat.toSet())
+        view.showPrice(seats.bookingPrice())
+        updateConfirmButtonState(limit)
+    }
```
위에 남겼던 대로 꼭 ArrayList 였어야 할까? 라는 생각이 들었습니다~🤔

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/seat/SeatRow.kt:
```diff
@@ -0,0 +1,44 @@
+import woowacourse.movie.domain.model.seat.Row
+import woowacourse.movie.domain.model.seat.Seat
+
+class SeatRow(
+    private val row: TableRow,
+    private val rowIndex: Int,
+    private val onSeatClick: (Seat) -> Unit,
+) {
```
해당 객체에서 View에 대한 의존성을 가지고 있군요..!
해당 객체의 ViewController 일까요? 도메인 객체일까요?🤔

chanho0908 replied:
SeatRow는 도메인 객체가 아닌 Activity에 TableLayout 관련한 코드가
너무 많아져서 별도의 객체로 분리한 객체입니다 ! ☺️

Gyuil-Hwnag replied:
그렇다면~ 네이밍을 좀 더 담당하는 기능에 맞도록 설정을 해보면 좋을 것 같다는 생각이 들었어요
히스토리를 모르는 사람이 해당 네이밍만 본다면 SeatRow는 도메인 객체로 볼 수도 있을 것 같다는 생각이 들어서요~
(사실 제가 그랬...어요😅)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/home/seat/SeatView.kt:
```diff
@@ -0,0 +1,31 @@
+import androidx.core.view.children
+import woowacourse.movie.domain.model.seat.Seat
+
+class SeatView(
+    private val seatTable: TableLayout,
+    private val onSeatClick: (Seat) -> Unit,
+) {
```
해당 객체는 SeatView의 책임을 담당하게 하기 위해서 만든 객체인가 보군요~

chanho0908 replied:
맞습니다 ! 테이블 레이아웃 관련한 코드가 너무 많아져서 액티비티가 뚠뚠해지더라구요 ,,,

Gyuil-Hwnag commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,33 @@
 <resources>
     <string name="app_name">Movie</string>
+    <string name="text_release_date">상영일: </string>
+    <string name="text_running_time">러닝타임: </string>
+    <string name="text_booking">지금 예매</string>
+    <string name="text_cancel_guide">영화 상영 시작 시간 15분 전까지\n취소가 가능합니다.</string>
+    <string name="text_decrease">-</string>
```
strings.xml 의 활용 너무 좋습니다💯

Gyuil-Hwnag commented on docs/feedback_1.md:
```diff
@@ -0,0 +1,99 @@
+## 3,4 단계 피드백 문서
+
+### 1. 리사이클러뷰의 리스너 인덱스 반환으로 인한 문제 해결 및 영화 데이터 접근 최적화
```
오호 피드백에 대한 내용을 정리하고 계시는군요~ 정말 최고네요💯

Gyuil-Hwnag reviewed (changes requested):
페토 안녕하세요~
해당 미션 구현 고생 많으셨습니다!
리뷰 반영하면서 구현한 내용 수정 너무 잘 해주셨습니다 💯

DataBinding에 대한 적용에 대해서 공부를 하신다음에 적용을 해주셨네요👍
현재 보기에 다른 Activity들도 적용을 해볼 수 있을 것 같다는 생각이 들었는데, 연습한다는 생각으로 적용해볼 수 있는 곳들에 적용을 해보시면서 손에 익도록 해보는 것도 좋겠다는 생각도 들었어요~
추가로 고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

chanho0908 committed:
feat : 좌석 선택 리스너 데이터 바인딩 적용

chanho0908 committed:
feat : 좌석 화면의 영화 제목 데이터 바인딩 적용

chanho0908 committed:
feat : 예매 완료 화면 데이터 바인딩 적용

chanho0908 committed:
feat : 포맷팅용 strings resource 정리

chanho0908 committed:
refactor : 영화 UI 모델 매핑 함수에 stringsResource 포맷팅 로직 추가

chanho0908 committed:
feat : layout name 속성으로 기본 프래그먼트 지정

chanho0908 commented on app/src/main/res/layout/activity_booking.xml:
```diff
@@ -72,7 +72,7 @@
                     <TextView
                         android:id="@+id/tv_screening_period"
-                        android:text="@{String.format(@string/formatter_date_period, model.screeningStartDate, model.screeningEndDate)}"
+                        android:text="@{model.screeningPeriod}"
```
데이터 바인딩을 사용하면 xml에서 직접 포맷팅을 할 수 있지만

상영 날짜의 경우 Localdate(2025-01-01) -> 날짜 형식 변경(2025.1.1) -> 날짜와 시간을 합친 문자열로 포매팅

총 두 번의 포맷팅을 거쳐야 하는데,

매 단계에서 포맷팅하는 코드가 너무 분산되는 것 같아 Ui모델을 만들 때 한번에 매핑하도록 수정했습니다.

chanho0908 commented:
안녕하세요 두루 ! 

저번 리뷰 요청 때 정말 많은 변경 사항이 있었기에 이번 리뷰 요청에성 변경 사항이 많지 않은 것 같네요 !

데이터 바인딩이 미적용 된 부분 위주로 수정했습니다.

다만 좌석을 선택할 때 마다 실시간으로 변경되는 티켓 가격의 경우
Prestener에서 매번 `showPrice(가격)` 을 호출하지 않고선

데이터 바인딩으로 실시간으로 적용해주기가 어려워 적용하지 못했습니다 🥲

chanho0908 commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -37,7 +37,6 @@ class MainActivity : AppCompatActivity() {
         val settingFragment = SettingFragment()
 
         if (savedInstanceState == null) {
-            replaceFragment(movieListFragment)
```
여기서 replace를 사용한 이유는 사실 강의 자료를 따라했던 것 같습니다 😓

프래그먼트는 액티비티가 파괴되면 프래그먼트 매니저에 의해 저장된다고 이해했습니다

때문에 저는 savedInstanceState가 null인 상태가 화면을 처음 켰을 때 라고 생각합니다.

replace는 기존의 프래그먼트를 제거하고 추가 (remove -> add) 하게되는데
그럼 저장되있는 프래그먼트가 없을텐데 왜 `replace`를 했지란 의문이 들었어요 !

이를 조사해본 결과 `android:name` 속성 같이 개발자가 실수로 프래그먼트를 추가했을 때
프래그먼트가 두 번 추가하는 실수를 방지하기 위한 안전 장치라고 생각했습니다

이런 실수를 누가 할까 싶지만 생각해보니 제가 그랬더라구요 ...
https://github.com/chanho0908/Today-I-Learned/blob/main/Fragment/Fragment_Duplicate_Add.md

이 내용에 대해서 동료 크루와 깊이 토론을 해봤는데 실제로 문제를 경험한 저는 이 이유에 대해 충분히 공감이 가지만
이런 경험이 없는 사람들에겐 공감하기 어려울 것 같아서 더 잘 설명해주고 싶습니다 !

제가 경험인 이유 왜에 다른 이유가 혹시 있을까요 ?

Gyuil-Hwnag replied:
흠~사실 저도 뭔가 예시가 없다보니, 정확히 어떤 부분에서 문제가 발생했을까? 라는 생각이 드는군요🤔
말씀해주신 내용이 아래와 같은 현상이 맞을까요?
1. `android:name="kr.techit.lion.search.SearchListFragment"`으로 설정
ㄴ 위에 처럼 설정했으니, 처음 시작은 SearchListFragment 가 된다.
2. add(R.id.fragmentContainerView, listFragment, ScreenStatus.List.name) 를 통해서 ListFragment 가 두개가 추가된다.
3. add(R.id.fragmentContainerView, mapFragment, ScreenStatus.Map.name) 를 통해서 MapFragment 가 추가됐다가 hide 를 하면서 스택은 아래와 같이 된다.

SearchList -> SearchList / SearchList -> SearchList / SearchList / Map -> SearchList / SearchList

chanho0908 replied:
앗 이건 제가 너무 어렵게 생각한 것 같네요 😅

savedInstanceState가 null == 상태가 화면을 처음 켰을 때라고 생각해 
저장되있는 프래그먼트가 없는 상태에서 왜 replace를 사용할까에 대한 의문이었습니다.

```
if (savedInstanceState == null) {
   supportFragmentManager.commit {
        replace(R.id.fragment_container_view, fragment)
    }
}
```

강의 자료 어디선가 어디선가 이 코드를 본 것 같은데 다시 보니까 없군요 ... 꿈을 꿨나 봅니다 😅

맨 처음 프래그먼트를 추가할 때 `add()` 또는 `android:name` 속성을 사용하면될 것 같습니다 !

Gyuil-Hwnag replied:
스스로 깨달음이 있으셨군요ㅎㅎ👍
정말 최고네요!💯

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/StringFormatter.kt:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    private const val DATE_TIME_DOT_FORMAT = "yyyy.M.d"
+    private const val THOUSAND_UNIT_FORMAT = "%,d"
```
이 내용과 관련이 있지는 않지만, 알아두면 혹시나 쓸모가 있을 수 있는 내용이라서 남겨두었습니다ㅎㅎ

얼마전에 서비스에서 크래시가 발생을 했어요.
- 요구사항 : 소수점 둘째짜리 까지 표기.
- 구현 : Float형의 경우 0.00f 로 지정.
- 크래시 : 0,00f 는 포맷을 하지 못한다.

원인 : 독일이나 유럽등의 나라에서는 0.00f 의 표기가 아닌 0,00f 로 표기가 들어간다고 하더군요!
그렇다면 이럴경우 해결방법은 어떤 것이 있을까요?

chanho0908 replied:
```
fun thousandFormat(price: Int): String {
        return THOUSAND_UNIT_FORMAT.format(Locale.GERMANY, price)
    }
```

format 메서드의 인자로 로케일을 전달할 수 있더군요 !

로케일 지정하거나 외부에서 전달받으면 어떨까 싶어요

Gyuil-Hwnag replied:
오호~ Locale을 외부에서 전달 받는 방법 좋습니다👍
Locale 을 지정하는 방법도 있긴 한데, 그렇게 되면 전세계를 기준으로 출시하는 경우 하나의 Locale 을 지정해야 한다는 점도 있게 될 것 같아요
DecimalFormat등을 활용해보는 것도 방법일 것 같긴 하네요ㅎㅎ

추후에 개발을 하다가 이런 이슈가 생겼을 때 기억속에서 어렴풋이 생각이 난다면 도움이 될 것 같아서 남겨둔 내용이었어요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/bindingadapter/SeatBindingAdapter.kt:
```diff
@@ -0,0 +1,25 @@
+package woowacourse.movie.view.bindingadapter
+
```
해당 BindingAdapter는 Seat 과 관련된 곳에서만 사용을 하는군요
아래의 코멘트를 참고한다면, 어떤 패키지로 이동을 해볼 수 있을까요~?🤔
이번 기회에 페토의 패키지를 나누는 기준을 한번 세워보고 가면 좋을 것 같아서요ㅎㅎ💪

[패키지 구조에 대한 코멘트](https://github.com/woowacourse/android-movie-theater/pull/102#discussion_r2073386608)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/handler/BookingActionHandler.kt:
```diff
@@ -0,0 +1,20 @@
+package woowacourse.movie.view.handler
+
```
BindingAdapters와 비슷한 내용입니다~
만약 프로젝트가 커져서 무수히 많은 Handler / BindingAdapters가 존재하게 된다면, 모든 파일들은 해당 패키지에 모이게 될까요??

chanho0908 replied:
이 부분도 패키징 수정 커밋에 답변 달겠습니다 !

Gyuil-Hwnag reviewed (changes requested):
페토 안녕하세요~
해당 미션 너무 고생 많으셨습니다!
리뷰 반영 정말 잘 해주셨네요.💯

하나의 파일에 많은 구현이 들어가는 것을 방지하기 위해서, 코드를 나누다보니 뭔가 구조가 복잡해졌다는 생각이 들었어요~
구현을 분리해주신 것은 좋으나, 자칫 잘못하면 프로젝트의 구조가 엄청 복잡해져서 히스토리를 파악하지 못한 상태로 본다면 해당 구조를 이해를 하는데에 많은 시간을 쏟을 수 있겠다는 생각이 들었어요.
다른 부분들은 너무 잘 구현을 해주셔서 이 부분에 대해서 한번 살짝 잡고 가면 좋을 것 같다는 생각이 들었어요.
도움이 될만한 코멘트 달았으니 확인해보시면 좋을 것 같습니다!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

Gyuil-Hwnag commented:
@chanho0908 
페토 안녕하세요~
이번 미션 정말 너무나 고생 많으셨어요!
다만, 뭔가 페토의 구현을 보다보니 이상하게 구조가 복잡하다는 느낌을 살짝 받게 되었어요🥲
개인적으로 이러한 느낌을 받았던 것은 아래의 이유였던 것 같아요..
1. 세분화된 패키지 구조 (패키지에 대한 정리)
2. ViewType 등 뎁스가 깊게 되어 있는 구현들
3. Handler 등 여러 객체의 인터페이스와 구현체들 이러한 구현체들과의 의존 관계

한번 페토와 같이 얘기를 해보면 좋을 주제인 것 같아서 이렇게 따로 내용으로 남겨두었습니다!
구현을 하게 된 계기와 충분히 많이 고민을 했던 내용들에 대해서는 저도 공감이 가는 내용들도 있었고, 너무 좋은 고민들이었어요👍

그러나 한편으로는 이러한 생각들도 들었어요.
`만약 제가 정말 같이 일하는 팀원이라면 무슨 생각이 들었을까?`
프로젝트는 한명이 만드는 것이 아니며, 여러 인원이 같이 만드는 서비스이므로, 저는 우선시 되어야 하는 것이 누가 보기에도 보기 쉽게 만드는 것이 중요하다는 생각이 들어요.

이런 것들을 위해서 디자인 패턴 및 아키텍쳐, 컨벤션등을 통일하는 것이기도 하고요.
아키텍처 및 디자인 패턴등은 여러 다른 이유들도 있겠지만, 저는 이것들의 사용으로서 최종적으로 장점은 `공통된 코드 형태형태가 나오고 그 누가 개발을 해도 일정한 품질을 유지하게 만들어 준다`는 생각이 들어요.
`이러한 디자인 패턴을 사용했고, 이러한 아키텍처로 구현했다.`이렇게 말하면 머리속에 아~ 어떠한 형태로 구현을 했겠구나 라는 생각이 드는 것처럼요.

그런 관점에서 봤을 때, 페토의 구현을 보면 너무 잘해주시긴 했었는데 뭔가 정재가 안된 것 같다는 찜찜함이 계속 남았던 것 같아요...
제가 실제로 페토가 구현하신 내용을 봤을 때도 여러 파일을 탐방하면서 해맸었기도 하네요🥲
이러한 부분들은 한번 저랑 같이 위의 생각들에 대해서 잡아보고 가면 좋을 것 같아서 이렇게 따로 내용을 남겨두었습니다.
(사실 파일 / 패키지만 정리해도 충분할 것 같긴 한데, 위에 남겨둔 이유들은 제가 보면서 찝찝했던 부분들을 모두 남겨두었던거라서 너무 신경은 안쓰셔도 될 것 같습니다!)

사실 `누가 보기에도 보기 쉽게 만드는 것이 중요하다는 생각` 이 부분은 사실 매번 저도 구현을 하면서 너무나 고민을 많이 하는 부분들이기도 합니다..

정작 저도 못하는 경우도 많았어요.. 그래서 구현된 것을 설명하느라 시간을 쏟고 그러는 경우도 많답니다..ㅎㅎ
저 뿐만 아니라 연차가 높으신 다른분들도 그러시는 경우도 있기도 하구요
그래도 머리속에 담아두고 하다보면 앞으로 페토가 개발을 하는데에 도움이 될 것 같다는 생각이 들어서요!

chanho0908 committed:
fix: 1~9일 날짜에 '0'이 붙어 '09일'처럼 표시되는 문제 수정

chanho0908 committed:
refactor: 도메인 모델 패키지 구조 변경

chanho0908 committed:
refactor: core 패키지 추가 및 util 함수 이동

chanho0908 committed:
refactor : home패키지에 묶인 화면들을 피쳐별로 분리

chanho0908 committed:
refactor : Domain -> UiModel, Recycleview Item class Mapper 함수를 클래스 파일로 분리

chanho0908 committed:
refacotr : Handler interface를 실제 사용하는 패키지로 분리

chanho0908 committed:
refactor : 리사이클러뷰 어댑터, 뷰홀더, 아이템 모델 패키지 통합

chanho0908 committed:
refactor : 좌석의 한 행을 담당하는 View Class의 네이밍 수정

chanho0908 committed:
refactor : BaseViewHolder의 bind 추상 메서드 제거

chanho0908 committed:
refactor : RecyclerView ViewType Emum class 위치를 어댑터 내부로 수정, 레이아웃 아이디 인자 제거

chanho0908 committed:
refactor : 직렬화 확장 함수 non-nullable (require) 로 수정

chanho0908 commented on app/src/main/java/woowacourse/movie/view/core/bindingadapter/ImageBindingAdapter.kt:
```diff
@@ -1,8 +1,8 @@
-package woowacourse.movie.view.bindingadapter
+package woowacourse.movie.view.core.bindingadapter
```
두루가 남겨주신 코멘트 링크를 달고 싶지만 코드레빗이 폭주해서 찾을수가 없네요 🥲

다음 미션에는 코드 레빗을 그냥 끄고 해보려 합니다... 너무 불편하네요 ...

> 프로젝트의 규모가 커져 Handler와 BindingAdapter가 무수히 많아지면 어떻게 관리할것인가 ?

위 질문에 대한 답변입니다

최우선적으로 하나의 피처에서만 사용되는 BindingAdapter나 Handler는 해당 피처 패키지에 위치시키며,

공용으로 사용되는 것들은 
+ core/bindingadapter 
+ core/handler 
로 사용될 것 같습니다 !

Gyuil-Hwnag replied:
core 로 해도 되고 common 등 여러 네이밍이 있을 것 같아요~
이건 페토만의 기준을 잘 세우시면 더욱더 좋을 것 같네요ㅎㅎ 

> 다음 미션에는 코드 레빗을 그냥 끄고 해보려 합니다... 너무 불편하네요 ...

저도...ㅎㅎ 뭔가 코멘트가 너무 무지막지하게 달리다보니 흐름이 끊겨서, 말씀드렸더니 이제부터는 커밋 할때마다 리뷰가 달리지 않도록 세팅을 변경해주셨다고 합니다!👍

chanho0908 commented on app/src/main/java/woowacourse/movie/view/movies/model/MovieRvItem.kt:
```diff
@@ -25,3 +29,20 @@ sealed class MovieRvItem(val viewType: ViewType) {
+    return AdItem(
+        imgResource = ImageSource.Resource(imgResource),
+    )
+}
```
도메인이 UiModel 또는 RecyclerView Item으로 매핑되는 과정을 바로 확인할 수 있다고 생각해서 같은 파일에 위치시켰습니다 !

chanho0908 commented on app/src/main/java/woowacourse/movie/view/movies/adapter/AdvertiseViewHolder.kt:
```diff
@@ -1,10 +1,10 @@
-package woowacourse.movie.view.movies.viewholder
+package woowacourse.movie.view.movies.adapter
```
기존에 adapter, viewholder로 분리된 패키지를 통합할지 분리할지 고민하면서
다른 크루들의 패키징도 조사해보고 이야기를 많이 나눠본 결과 대부분
adapter와 viewHolder를 통합한단 의견이 많아서 합치게 되었습니다 ! ☺️

Gyuil-Hwnag replied:
아주 좋습니다ㅎㅎ
개인적으로도 사용성이 비슷한것은 같이 묶으려고 하는편이기는 했어요~👍

chanho0908 commented on app/src/main/java/woowacourse/movie/view/core/base/BaseViewHolder.kt:
```diff
@@ -5,11 +5,9 @@ import android.view.ViewGroup
-        LayoutInflater.from(parent.context).inflate(layoutRes, parent, false),
-    ) {
-    abstract fun bind(item: T)
-}
```
앞서 말씀드린 의견을 고민해본 결과 `bind` 함수만으로 모든 바인딩 로직을 담을 수 없다고 생각해서 제거했습니다 !

chanho0908 commented on app/src/main/java/woowacourse/movie/view/movies/adapter/MovieAdapter.kt:
```diff
@@ -43,5 +45,10 @@ class MovieAdapter(
         }
     }
 
+    enum class ViewType {
+        VIEW_TYPE_ADVERTISEMENT,
+        VIEW_TYPE_MOVIE,
+    }
```
두 가지 변경사항이 있습니다 !

- `ViewType class`를 `Adapter` 내부로 이동
   - `Adapter`에서` ViewType`을 찾기 위한 노력을 안해도 된다 !
- `ViewType` 이 갖고 있던 Layout Id 제거
  - 안드로이드 의존성 제거
  - `Adapter Class` 내에서 바로 xml로 이동 가능

해당 이유들에 의해서 수정해봤습니다☺️

디노의 코드를 제 것으로 만드는 과정이었던 것 같아요 
개인적으로 만족할만한 방식이라고 생각되서 앞으로도 이 방식을 사용할 것 같네요 !

Gyuil-Hwnag replied:
이렇게 한걸음 발전해나가고 페토의 기준을 세워가는군요💯

chanho0908 reviewed (commented):
안녕하세요 두루 !
저번 미션에서도, 이번 미션에서도 두루께서 정말 많은 가르침을 주셔서 정말 감사합니다 ☺️

가뜩이나 코드레빗이 폭주하면서 리뷰하기 힘드셨을 것 같은데
제 어지러운 패키지 구조까지 더해져 리뷰가 너무 힘드셨을 것 같아 죄송하네요 🥲

두루께서 괜찮으시다면 다음 3,4단계는 코드 레빗을 끄고 진행하면 어떨까 싶어요 ! 
두루의 소중한 리뷰를 찾기 힘든 것 같습니다 😅

두루께서 남겨주신 장문의 편지가 제가 고민하던 내용인 것 같아요 ! 🤔

얼마전 제임스와의 원온원에서 드렸던 질문 중 하나가 바로 요겁니다
* 함께하기 좋은 크루란 어떤 것일까? 

이 질문을 하게된 계기는 우테코에 처음 와서 레벨 1에선 약간의 적응과 우테코가 추구하는 방향을 알아가는 단계였다면,

레벨 2에선 일련의 일들을 겪고 다음 질문들이 떠올랐어요
+ 나는 과연 좋은 사람인가?
+ 나는 함께 일하기 좋은 동료인가? 
+ 함께 일하기 좋은 동료란 무엇일까 

같은 질문들이 스스로 떠오르기 시작했어요 !

고민해본 결과 이건 제가 앞으로 더 많은 경험을 쌓아가며 스스로 답을 찾아가야 하는 것 같습니다 😉

저는 제 스스로 메타인지가 굉장히 잘 되어있다고 생각해서 스스로 고쳐야할 부분들을 항상 느끼곤 한답니다 하하하

부끄럽지만 썰을 하나 풀자면 우테코 생활을 해가면서 점점 피곤에 쩌들어 힘들 때 때 크루들에게 다소 예민하게 대한다는 것을 느꼈어요

제 기분이 태도가 안된다고 생각해 고치려 노력하고 있습니다
크루들이 어떻게 생각하실진 모르겠네요 하하하하하하 😂

조금 이야기가 샌 것 같은데, 다시 돌아와서 !

놀랍게도 두루뿐만 아니라 다른 크루분도 제 코드를 보실 때 너무 찾기 힘들다고 하셨습니다 🥲🥲

그래서 패키징을 수정할 때 다른 크루들의 모두 뜯어보며 패키징을 어떻게 하셨는지 분석했어요 !
다른 크루와 의견도 나눠보면서 많은 고민을 했던 것 같습니다.

돌아보니 제가 간과한 부분이 있었던 것 같습니다

리뷰어님들도 이 미션을 함께 해결해나가는 "동료" 란 것을 제가 몰랐던 것 같아요

남은 3,4단계 미션에선 항상 두루도 이 미션을 함께 하는 동료란 것을 생각하면서
최대한 보기 쉽게 만들어 보려고 노력해보겠습니다 😄

항상 바쁘신 와중에도 왕왕 길게 의견 남겨주셔서 감사합니다 두루 ! ☺️

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/booking/BookingActivityTest.kt:
```diff
@@ -0,0 +1,108 @@
+        onView(withId(R.id.tv_people_count)).check(matches(withText("1")))
+    }
+
+    @Test
+    fun `화면이_회전_되어도_인원수가_유지된다`() {
+        // given
+        onView(withId(R.id.btn_increase)).perform(click())
+
+        // When
+        scenario.onActivity { activity ->
+            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
+        }
+
+        // then
+        onView(withId(R.id.tv_people_count)).check(matches(withText("2")))
```
화면 전환에 대한 테스트도 정말 잘 작성해주셨네요👍

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/complete/BookingCompleteActivityTest.kt:
```diff
@@ -0,0 +1,60 @@
+        ActivityScenario.launch<BookingCompleteActivity>(intent)
+    }
+
+    @Test
+    fun `전달_받은_영화_이름_상영_시간_예매_인원_예매_가격을_출력한다`() {
```
이 부분은 각각으로 나눠서 테스트를 해보는 것도 좋겠다는 생각이 들었어요~
이유는 세분화한다면, 테스트가 실패했을 때 어떤 부분에서 실패를 하는지 좀 더 한눈에 알아볼 수 있다는 장점도 있겠어요

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/seat/SeatActivityTest.kt:
```diff
@@ -0,0 +1,70 @@
+        onView(withId(R.id.tv_price)).check(matches(withText("20,000원")))
+    }
+
+    @Test
+    fun `좌석을_선택_해제하면_감소된_좌석_가격을_출력한다`() {
+        // given
+        onView(withId(R.id.tv_price)).check(matches(withText("0원")))
+
+        // when
+        onView(withId(R.id.a1)).perform(click())
+        onView(withId(R.id.a1)).perform(click())
+
+        // then
+        onView(withId(R.id.tv_price)).check(matches(withText("0원")))
+    }
```
중간에 올랐던 가격에 대한 체크까지 있다면, 이중으로 테스트가 가능해지겠군요~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/domain/model/seat/SeatPolicy.kt:
```diff
@@ -0,0 +1,26 @@
+package woowacourse.movie.domain.model.seat
+
+sealed interface SeatPolicy {
+    val price: Int
```
아래와 같이 enum class를 활용하는 방법도 있을 것 같은데, sealed class를 사용해주셨던 이유가 있을까요~🤔
~~~kotlin
enum class SeatPolicy(val price: Int) {
    S(15_000),
    B(10_000),
    A(12_000);
       
    companion object {
        fun from(row: Int): SeatPolicy {
            when (row) {
                1, 2 -> S
                3, 4 -> B
                else -> A
            }
        }
    }
}
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -0,0 +1,77 @@
+        movieListFragment: MovieListFragment,
+        settingFragment: SettingFragment,
+    ) {
+        binding.navigation.setOnItemSelectedListener { item ->
+            when (item.itemId) {
+                R.id.action_history -> {
+                    replaceFragment(historyFragment)
+                    true
+                }
+
+                R.id.action_home -> {
+                    replaceFragment(movieListFragment)
+                    true
+                }
+
+                R.id.action_setting -> {
+                    replaceFragment(settingFragment)
+                    true
+                }
+
+                else -> return@setOnItemSelectedListener false
+            }
+        }
```
아래와 같이 작성을 한다면, 중복된 리턴에 대해서 줄여볼 수 있겠다는 생각도 듭니다~
~~~kotlin
        binding.navigation.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.action_history -> replaceFragment(historyFragment)
                R.id.action_home -> replaceFragment(movieListFragment)
                R.id.action_setting -> replaceFragment(settingFragment)
                else -> return@setOnItemSelectedListener false
            }
            return@setOnItemSelectedListener true
        }
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/domain/model/Booking.kt:
```diff
@@ -0,0 +1,44 @@
+
+    fun decreasePeopleCount() = this.copy(count = count.decrease())
+
+    fun restore(
+        bookingDate: LocalDate,
+        bookingTime: LocalTime,
+        peopleCount: Int,
+    ): Booking {
+        val newCount = count.modify(peopleCount)
+        return this.copy(bookingDate = bookingDate, bookingTime = bookingTime, count = newCount)
+    }
```
restore 에서 copy를 사용하게 된 생각의 흐름이 궁금합니다~

Gyuil-Hwnag replied:
나아가서는 깊은 복사 / 얕은 복사까지 살펴볼만한 키워드들이 나오겠군요!💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/movies/adapter/AdvertiseViewHolder.kt:
```diff
@@ -0,0 +1,18 @@
+import woowacourse.movie.view.core.base.BaseViewHolder
+import woowacourse.movie.view.movies.adapter.model.MovieRvItem
+
+class AdvertiseViewHolder(
+    parent: ViewGroup,
```
ViewHolder 객체를 Adapter 내에 넣어볼 수도 있을 것 같은데, 이렇게 하면 어떤 차이점이 존재하게 될까요?🤔
inner class등에 대한 키워드를 한번 살펴볼 수 있는 키워드일 것 같아서요!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/uiModel/MovieUiModel.kt:
```diff
@@ -0,0 +1,30 @@
+import woowacourse.movie.view.core.bindingadapter.ImageSource
+import woowacourse.movie.view.core.util.StringFormatter
+
+data class MovieUiModel(
+    val title: String,
+    val posterResource: ImageSource,
+    val screeningPeriod: String,
+    val runningTime: String,
+)
```
UiModel 에 대해서는 페로로도 잡담채널에 올려주신 만큼 한번 더 생각을 해보면 좋을 것 같다는 생각도 들긴 하네요~💪

Gyuil-Hwnag commented on app/src/test/java/woowacourse/movie/presenter/BookingPresenterTest.kt:
```diff
@@ -0,0 +1,151 @@
+import java.time.LocalDate
+import java.time.LocalDateTime
+
+class BookingPresenterTest {
+    private val view: BookingContract.View = mockk<BookingContract.View>(relaxed = true)
+    private val movieStore = mockk<MovieStore>()
```
Presenter에 대한 테스트와 mockk의 활용 너무 좋네요👍
@BeforeEach로 세팅을 해주는 만큼 @AfterEach 어노테이션을 활용해볼 수도 있겠다는 생각도 드네요~

Gyuil-Hwnag reviewed (approved):
페토 안녕하세요~
해당 미션 너무 고생 많으셨습니다!
리뷰 반영 정말 잘 해주셨네요💯
해당 미션은 다음 미션 진행을 위하여 머지하도록 하겠습니다.

패키지 및 전반적인 구현에 대한 리팩토링 너무 잘해주셨어요💯💯💯
일정도 많이 빠듯하고, 시간 압박등 여러가지로 신경쓰이는 것들이 많을텐데 이럴 때는 가끔 여러가지로 스트레스가 쌓일 때는 잠깐이라도 노트북을 덮고 있어보는 것도 많은 도움이 되긴 하더라구요
이러한 작은 행동들이 꾸준히 할 수 있도록 도움을 주는 것 같아요
작은 방법들이라도 페토만의 스트레스를 해소하는 것들을 잘 찾아내 나가시길 바랄게요!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪
언제나 응원하겠습니다!

Gyuil-Hwnag commented:
@chanho0908 
페토 안녕하세요~
이번 미션 정말 너무 고생 많으셨어요!

개인적으로 리뷰로만 보긴 했지만, 페토는 정말 좋은 동료라고 생각이 들어요
이전 미션에서 열정적인 모습들을 많이 보여주셨어서, 리뷰하는 저의 입장에서도 페토의 이러한 열정적인 모습들 덕분에 힘이나고 재밌게 리뷰를 했던 것 같아요!

코멘트로도 남겨두긴 했지만, 가끔 이러한 압박들이 있을 때 잘 해소할 수 있는 것들을 잘 찾으시길 바랄게요!
(저도 스트레스 푸는 것들이 없었다면 꾸준히 해나가지 못했을 거에요)
> 이럴 때는 가끔 여러가지로 스트레스가 쌓일 때는 잠깐이라도 노트북을 덮고 있어보는 것도 많은 도움이 되긴 하더라구요
이러한 작은 행동들이 꾸준히 할 수 있도록 도움을 주는 것 같아요
작은 방법들이라도 페토만의 스트레스를 해소하는 것들을 잘 찾아내 나가시길 바랄게요!

개인적으로 회사에서 혼자서 개발을 하던 시절도 있다보니, 같은 것을 바라보고 같이 고민을 해볼 수 있는 동료들이 있다는 것에 대한 감사함이 정말 큰 것 같아요.
(저도 리뷰를 하면서 리뷰이분들한테 힘도 얻고 여러가지 아이디어도 얻어가고 있는 것 같아요)

말씀해주신대로, 리뷰어와 리뷰이를 떠나서 이번 미션을 같이 진행해나가는 하나의 동료로서 이번 미션도 잘 진행해나가보도록 해봐요ㅎㅎ
이번 미션 정말 너무나 고생 많으셨습니다!
다음 미션도 같이 화이팅해보도록 합시다💪
